// @jubbio/core/dist/sharding/ShardingManager.js
"use strict";
/**
 * ShardingManager - Multi-process bot support
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShardClientUtil = exports.ShardingManager = exports.Shard = exports.ShardStatus = void 0;
const events_1 = require("events");
const child_process_1 = require("child_process");
const path_1 = __importDefault(require("path"));
/**
 * Shard status
 */
var ShardStatus;
(function (ShardStatus) {
    ShardStatus[ShardStatus["Ready"] = 0] = "Ready";
    ShardStatus[ShardStatus["Connecting"] = 1] = "Connecting";
    ShardStatus[ShardStatus["Reconnecting"] = 2] = "Reconnecting";
    ShardStatus[ShardStatus["Idle"] = 3] = "Idle";
    ShardStatus[ShardStatus["Nearly"] = 4] = "Nearly";
    ShardStatus[ShardStatus["Disconnected"] = 5] = "Disconnected";
    ShardStatus[ShardStatus["WaitingForGuilds"] = 6] = "WaitingForGuilds";
    ShardStatus[ShardStatus["Identifying"] = 7] = "Identifying";
    ShardStatus[ShardStatus["Resuming"] = 8] = "Resuming";
})(ShardStatus || (exports.ShardStatus = ShardStatus = {}));
/**
 * Represents a single shard
 */
class Shard extends events_1.EventEmitter {
    /** The manager that spawned this shard */
    manager;
    /** The shard ID */
    id;
    /** The child process */
    process = null;
    /** Whether the shard is ready */
    ready = false;
    /** Shard status */
    status = ShardStatus.Idle;
    /** Environment variables for the shard */
    env;
    constructor(manager, id) {
        super();
        this.manager = manager;
        this.id = id;
        this.env = {
            SHARD_ID: String(id),
            SHARD_COUNT: String(manager.totalShards),
            JUBBIO_TOKEN: manager.token ?? '',
        };
    }
    /**
     * Spawn the shard process
     */
    async spawn(timeout = 30000) {
        if (this.process) {
            throw new Error(`Shard ${this.id} already has a process`);
        }
        this.status = ShardStatus.Connecting;
        this.process = (0, child_process_1.fork)(this.manager.file, this.manager.shardArgs, {
            env: { ...process.env, ...this.env },
            execArgv: this.manager.execArgv,
        });
        this.process.on('message', this._handleMessage.bind(this));
        this.process.on('exit', this._handleExit.bind(this));
        this.process.on('error', this._handleError.bind(this));
        // Wait for ready
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                reject(new Error(`Shard ${this.id} took too long to become ready`));
            }, timeout);
            this.once('ready', () => {
                clearTimeout(timer);
                resolve(this.process);
            });
            this.once('disconnect', () => {
                clearTimeout(timer);
                reject(new Error(`Shard ${this.id} disconnected before becoming ready`));
            });
        });
    }
    /**
     * Kill the shard process
     */
    kill() {
        if (this.process) {
            this.process.removeAllListeners();
            this.process.kill();
            this.process = null;
        }
        this.status = ShardStatus.Disconnected;
        this.ready = false;
    }
    /**
     * Respawn the shard
     */
    async respawn(options) {
        this.kill();
        if (options?.delay) {
            await new Promise(r => setTimeout(r, options.delay));
        }
        return this.spawn(options?.timeout);
    }
    /**
     * Send a message to the shard
     */
    send(message) {
        return new Promise((resolve, reject) => {
            if (!this.process) {
                reject(new Error(`Shard ${this.id} has no process`));
                return;
            }
            this.process.send(message, (err) => {
                if (err)
                    reject(err);
                else
                    resolve();
            });
        });
    }
    /**
     * Evaluate code on the shard
     */
    async eval(script) {
        const _eval = typeof script === 'function' ? `(${script})(this)` : script;
        return new Promise((resolve, reject) => {
            const id = Date.now().toString(36) + Math.random().toString(36);
            const handler = (message) => {
                if (message._evalId !== id)
                    return;
                this.process?.off('message', handler);
                if (message._error) {
                    reject(new Error(message._error));
                }
                else {
                    resolve(message._result);
                }
            };
            this.process?.on('message', handler);
            this.send({ _eval, _evalId: id }).catch(reject);
        });
    }
    /**
     * Fetch a client property
     */
    async fetchClientValue(prop) {
        return this.eval(`this.${prop}`);
    }
    _handleMessage(message) {
        if (message._ready) {
            this.ready = true;
            this.status = ShardStatus.Ready;
            this.emit('ready');
            this.manager.emit('shardReady', this.id);
            return;
        }
        if (message._disconnect) {
            this.ready = false;
            this.status = ShardStatus.Disconnected;
            this.emit('disconnect');
            this.manager.emit('shardDisconnect', this.id);
            return;
        }
        if (message._reconnecting) {
            this.ready = false;
            this.status = ShardStatus.Reconnecting;
            this.emit('reconnecting');
            this.manager.emit('shardReconnecting', this.id);
            return;
        }
        this.emit('message', message);
        this.manager.emit('message', this.id, message);
    }
    _handleExit(code, signal) {
        this.ready = false;
        this.status = ShardStatus.Disconnected;
        this.process = null;
        this.emit('death', { code, signal });
        this.manager.emit('shardDeath', this.id, { code, signal });
        if (this.manager.respawn) {
            this.spawn().catch(err => {
                this.manager.emit('shardError', this.id, err);
            });
        }
    }
    _handleError(error) {
        this.emit('error', error);
        this.manager.emit('shardError', this.id, error);
    }
}
exports.Shard = Shard;
/**
 * Manages multiple shards for large bots
 */
class ShardingManager extends events_1.EventEmitter {
    /** Path to the bot file */
    file;
    /** Total number of shards */
    totalShards;
    /** List of shard IDs to spawn */
    shardList;
    /** Sharding mode */
    mode;
    /** Whether to respawn shards */
    respawn;
    /** Arguments to pass to shards */
    shardArgs;
    /** Arguments to pass to node */
    execArgv;
    /** Bot token */
    token;
    /** Collection of shards */
    shards = new Map();
    constructor(file, options = {}) {
        super();
        this.file = path_1.default.resolve(file);
        this.totalShards = options.totalShards ?? 'auto';
        this.shardList = options.shardList === 'auto' ? [] : (options.shardList ?? []);
        this.mode = options.mode ?? 'process';
        this.respawn = options.respawn ?? true;
        this.shardArgs = options.shardArgs ?? [];
        this.execArgv = options.execArgv ?? [];
        this.token = options.token;
    }
    /**
     * Spawn all shards
     */
    async spawn(options) {
        // Determine shard count
        if (this.totalShards === 'auto' || options?.amount === 'auto') {
            this.totalShards = await this.fetchRecommendedShards();
        }
        else if (options?.amount) {
            this.totalShards = options.amount;
        }
        // Build shard list if not specified
        if (this.shardList.length === 0) {
            this.shardList = Array.from({ length: this.totalShards }, (_, i) => i);
        }
        // Spawn shards sequentially with delay
        const delay = options?.delay ?? 5500;
        for (const id of this.shardList) {
            const shard = this.createShard(id);
            await shard.spawn(options?.timeout);
            if (id !== this.shardList[this.shardList.length - 1]) {
                await new Promise(r => setTimeout(r, delay));
            }
        }
        return this.shards;
    }
    /**
     * Create a shard
     */
    createShard(id) {
        const shard = new Shard(this, id);
        this.shards.set(id, shard);
        return shard;
    }
    /**
     * Fetch recommended shard count from API
     */
    async fetchRecommendedShards() {
        // In a real implementation, this would call the API
        // For now, return a default
        return 1;
    }
    /**
     * Broadcast a message to all shards
     */
    async broadcast(message) {
        const promises = [...this.shards.values()].map(shard => shard.send(message));
        return Promise.all(promises);
    }
    /**
     * Broadcast an eval to all shards
     */
    async broadcastEval(script) {
        const promises = [...this.shards.values()].map(shard => shard.eval(script));
        return Promise.all(promises);
    }
    /**
     * Fetch a client value from all shards
     */
    async fetchClientValues(prop) {
        return this.broadcastEval(`this.${prop}`);
    }
    /**
     * Respawn all shards
     */
    async respawnAll(options) {
        for (const shard of this.shards.values()) {
            await shard.respawn({ delay: options?.respawnDelay, timeout: options?.timeout });
            if (options?.shardDelay) {
                await new Promise(r => setTimeout(r, options.shardDelay));
            }
        }
        return this.shards;
    }
}
exports.ShardingManager = ShardingManager;
/**
 * Shard client utilities - use in bot file
 */
class ShardClientUtil {
    /** The client */
    client;
    /** The shard ID */
    id;
    /** Total shard count */
    count;
    constructor(client) {
        this.client = client;
        this.id = parseInt(process.env.SHARD_ID ?? '0', 10);
        this.count = parseInt(process.env.SHARD_COUNT ?? '1', 10);
    }
    /**
     * Send a message to the parent process
     */
    send(message) {
        return new Promise((resolve, reject) => {
            process.send?.(message, (err) => {
                if (err)
                    reject(err);
                else
                    resolve();
            });
        });
    }
    /**
     * Fetch a client value from all shards
     */
    async fetchClientValues(prop) {
        return this.broadcastEval(`this.${prop}`);
    }
    /**
     * Broadcast an eval to all shards
     */
    async broadcastEval(script) {
        const _eval = typeof script === 'function' ? `(${script})(this)` : script;
        return new Promise((resolve, reject) => {
            const id = Date.now().toString(36) + Math.random().toString(36);
            const handler = (message) => {
                if (message._broadcastEvalId !== id)
                    return;
                process.off('message', handler);
                if (message._error) {
                    reject(new Error(message._error));
                }
                else {
                    resolve(message._results);
                }
            };
            process.on('message', handler);
            this.send({ _broadcastEval: _eval, _broadcastEvalId: id }).catch(reject);
        });
    }
    /**
     * Signal ready to the parent process
     */
    ready() {
        process.send?.({ _ready: true });
    }
    /**
     * Get the shard ID for a guild
     */
    static shardIdForGuildId(guildId, shardCount) {
        const id = BigInt(guildId);
        return Number(id >> 22n) % shardCount;
    }
}
exports.ShardClientUtil = ShardClientUtil;
exports.default = ShardingManager;
