// @jubbio/core/dist/utils/BitField.js
"use strict";
/**
 * Generic BitField class
 * API compatible with Discord.js BitField
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitField = void 0;
/**
 * Data structure for bit fields
 */
class BitField {
    /** The raw bits */
    bitfield;
    /** Flags for this bitfield (override in subclass) */
    static Flags = {};
    /** Default bit value */
    static DefaultBit = 0n;
    constructor(bits = BitField.DefaultBit) {
        this.bitfield = this.constructor.resolve(bits);
    }
    /**
     * Check if this bitfield has a bit
     */
    has(bit) {
        const resolved = this.constructor.resolve(bit);
        return (this.bitfield & resolved) === resolved;
    }
    /**
     * Check if this bitfield has any of the bits
     */
    any(bits) {
        const resolved = this.constructor.resolve(bits);
        return (this.bitfield & resolved) !== 0n;
    }
    /**
     * Add bits to this bitfield
     */
    add(...bits) {
        let total = 0n;
        for (const bit of bits) {
            total |= this.constructor.resolve(bit);
        }
        this.bitfield |= total;
        return this;
    }
    /**
     * Remove bits from this bitfield
     */
    remove(...bits) {
        let total = 0n;
        for (const bit of bits) {
            total |= this.constructor.resolve(bit);
        }
        this.bitfield &= ~total;
        return this;
    }
    /**
     * Serialize to array of flag names
     */
    toArray() {
        const flags = this.constructor.Flags;
        const result = [];
        for (const [name, bit] of Object.entries(flags)) {
            if (this.bitfield & bit) {
                result.push(name);
            }
        }
        return result;
    }
    /**
     * Serialize to JSON
     */
    toJSON() {
        return typeof this.bitfield === 'bigint'
            ? this.bitfield.toString()
            : this.bitfield;
    }
    /**
     * Get string representation
     */
    toString() {
        return this.bitfield.toString();
    }
    /**
     * Get iterator
     */
    *[Symbol.iterator]() {
        yield* this.toArray();
    }
    /**
     * Freeze this bitfield
     */
    freeze() {
        return Object.freeze(this);
    }
    /**
     * Check equality
     */
    equals(other) {
        return this.bitfield === this.constructor.resolve(other);
    }
    /**
     * Clone this bitfield
     */
    clone() {
        return new this.constructor(this.bitfield);
    }
    /**
     * Resolve a bit to the numeric type
     */
    static resolve(bit) {
        if (typeof bit === 'bigint' || typeof bit === 'number') {
            return bit;
        }
        if (bit instanceof BitField) {
            return bit.bitfield;
        }
        if (typeof bit === 'string') {
            const resolved = this.Flags[bit];
            if (resolved === undefined) {
                throw new Error(`Unknown bit: ${bit}`);
            }
            return resolved;
        }
        if (Array.isArray(bit)) {
            let result = this.DefaultBit;
            for (const b of bit) {
                const resolved = this.resolve(b);
                result |= resolved;
            }
            return result;
        }
        throw new Error(`Invalid bit: ${bit}`);
    }
}
exports.BitField = BitField;
exports.default = BitField;
