// @jubbio/core/dist/utils/Collector.js
"use strict";
/**
 * Collectors for awaiting messages, reactions, and interactions
 * API compatible with Discord.js Collectors
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactionCollector = exports.InteractionCollector = exports.MessageCollector = exports.Collector = void 0;
exports.awaitMessages = awaitMessages;
exports.awaitReactions = awaitReactions;
const events_1 = require("events");
const Collection_1 = require("./Collection");
/**
 * Abstract base class for collectors
 */
class Collector extends events_1.EventEmitter {
    /** The client that instantiated this collector */
    client;
    /** The items collected */
    collected = new Collection_1.Collection();
    /** Whether the collector has ended */
    ended = false;
    /** The reason the collector ended */
    endReason = null;
    /** Filter function */
    filter;
    /** Collector options */
    options;
    /** Number of items processed */
    _processedCount = 0;
    /** Timeout for time limit */
    _timeout = null;
    /** Timeout for idle limit */
    _idleTimeout = null;
    constructor(client, options = {}) {
        super();
        this.client = client;
        this.options = options;
        this.filter = options.filter ?? (() => true);
        this.handleCollect = this.handleCollect.bind(this);
        this.handleDispose = this.handleDispose.bind(this);
        if (options.time) {
            this._timeout = setTimeout(() => this.stop('time'), options.time);
        }
        if (options.idle) {
            this._idleTimeout = setTimeout(() => this.stop('idle'), options.idle);
        }
    }
    /**
     * Handle an item being collected
     */
    async handleCollect(item) {
        if (this.ended)
            return;
        this._processedCount++;
        const filterResult = await this.filter(item, this.collected);
        if (!filterResult)
            return;
        const key = this.collect(item);
        if (key === null)
            return;
        this.collected.set(key, item);
        this.emit('collect', item);
        // Reset idle timer
        if (this._idleTimeout) {
            clearTimeout(this._idleTimeout);
            this._idleTimeout = setTimeout(() => this.stop('idle'), this.options.idle);
        }
        // Check limits
        if (this.options.max && this.collected.size >= this.options.max) {
            this.stop('limit');
        }
        if (this.options.maxProcessed && this._processedCount >= this.options.maxProcessed) {
            this.stop('processedLimit');
        }
    }
    /**
     * Handle an item being disposed
     */
    handleDispose(item) {
        if (!this.options.dispose)
            return;
        const key = this.dispose(item);
        if (key === null)
            return;
        if (this.collected.has(key)) {
            this.collected.delete(key);
            this.emit('dispose', item);
        }
    }
    /**
     * Stop the collector
     */
    stop(reason = 'user') {
        if (this.ended)
            return;
        this.ended = true;
        this.endReason = reason;
        if (this._timeout) {
            clearTimeout(this._timeout);
            this._timeout = null;
        }
        if (this._idleTimeout) {
            clearTimeout(this._idleTimeout);
            this._idleTimeout = null;
        }
        this.emit('end', this.collected, reason);
    }
    /**
     * Reset the collector's timer
     */
    resetTimer(options = {}) {
        if (this._timeout) {
            clearTimeout(this._timeout);
            this._timeout = null;
        }
        if (this._idleTimeout) {
            clearTimeout(this._idleTimeout);
            this._idleTimeout = null;
        }
        if (options.time ?? this.options.time) {
            this._timeout = setTimeout(() => this.stop('time'), options.time ?? this.options.time);
        }
        if (options.idle ?? this.options.idle) {
            this._idleTimeout = setTimeout(() => this.stop('idle'), options.idle ?? this.options.idle);
        }
    }
    /**
     * Check the end conditions
     */
    checkEnd() {
        const reason = this.endReason;
        if (reason) {
            this.stop(reason);
            return true;
        }
        return false;
    }
    /**
     * Get the next item
     */
    get next() {
        return new Promise((resolve, reject) => {
            if (this.ended) {
                reject(new Error('Collector has ended'));
                return;
            }
            const cleanup = () => {
                this.removeListener('collect', onCollect);
                this.removeListener('end', onEnd);
            };
            const onCollect = (item) => {
                cleanup();
                resolve(item);
            };
            const onEnd = () => {
                cleanup();
                reject(new Error('Collector ended'));
            };
            this.on('collect', onCollect);
            this.on('end', onEnd);
        });
    }
}
exports.Collector = Collector;
/**
 * Collector for messages
 */
class MessageCollector extends Collector {
    channelId;
    messageHandler;
    constructor(client, channelId, options = {}) {
        super(client, options);
        this.channelId = channelId;
        // Get bot's user ID to filter out bot's own messages
        const botUserId = client.user?.id;
        this.messageHandler = (message) => {
            if (message.channel_id === this.channelId || message.channelId === this.channelId) {
                // Automatically filter out bot's own messages
                const authorId = message.author?.id || message.author_id;
                if (botUserId && authorId === botUserId)
                    return;
                // Also filter by application_id (bot messages have this)
                if (message.application_id || message.applicationId)
                    return;
                this.handleCollect(message);
            }
        };
        client.on('messageCreate', this.messageHandler);
        this.once('end', () => {
            client.removeListener('messageCreate', this.messageHandler);
        });
    }
    collect(message) {
        return message.id ?? null;
    }
    dispose(message) {
        return message.id ?? null;
    }
}
exports.MessageCollector = MessageCollector;
/**
 * Collector for interactions (buttons, select menus, etc.)
 */
class InteractionCollector extends Collector {
    channelId;
    guildId;
    messageId;
    interactionType;
    componentType;
    interactionHandler;
    constructor(client, options = {}) {
        super(client, options);
        this.channelId = options.channelId;
        this.guildId = options.guildId;
        this.messageId = options.messageId;
        this.interactionType = options.interactionType
            ? Array.isArray(options.interactionType) ? options.interactionType : [options.interactionType]
            : undefined;
        this.componentType = options.componentType
            ? Array.isArray(options.componentType) ? options.componentType : [options.componentType]
            : undefined;
        this.interactionHandler = (interaction) => {
            // Filter by channel
            if (this.channelId && interaction.channelId !== this.channelId)
                return;
            // Filter by guild
            if (this.guildId && interaction.guildId !== this.guildId)
                return;
            // Filter by message
            if (this.messageId && interaction.message?.id !== this.messageId)
                return;
            // Filter by interaction type
            if (this.interactionType && !this.interactionType.includes(interaction.type))
                return;
            // Filter by component type
            if (this.componentType && interaction.componentType && !this.componentType.includes(interaction.componentType))
                return;
            this.handleCollect(interaction);
        };
        client.on('interactionCreate', this.interactionHandler);
        this.once('end', () => {
            client.removeListener('interactionCreate', this.interactionHandler);
        });
    }
    collect(interaction) {
        return interaction.id ?? null;
    }
    dispose(interaction) {
        return interaction.id ?? null;
    }
}
exports.InteractionCollector = InteractionCollector;
/**
 * Collector for reactions
 */
class ReactionCollector extends Collector {
    messageId;
    reactionHandler;
    constructor(client, messageId, options) {
        super(client, options);
        this.messageId = messageId;
        this.reactionHandler = (reaction) => {
            if (reaction.message_id === this.messageId || reaction.messageId === this.messageId) {
                this.handleCollect(reaction);
            }
        };
        client.on('messageReactionAdd', this.reactionHandler);
        this.once('end', () => {
            client.removeListener('messageReactionAdd', this.reactionHandler);
        });
    }
    collect(reaction) {
        // Key is emoji identifier
        return reaction.emoji?.id ?? reaction.emoji?.name ?? null;
    }
    dispose(reaction) {
        return reaction.emoji?.id ?? reaction.emoji?.name ?? null;
    }
}
exports.ReactionCollector = ReactionCollector;
/**
 * Await messages helper
 */
function awaitMessages(client, channelId, options = {}) {
    return new Promise((resolve, reject) => {
        const collector = new MessageCollector(client, channelId, options);
        collector.once('end', (collected, reason) => {
            if (options.max && collected.size < options.max) {
                reject(new Error(`Collector ended with reason: ${reason}`));
            }
            else {
                resolve(collected);
            }
        });
    });
}
/**
 * Await reactions helper
 */
function awaitReactions(client, messageId, options) {
    return new Promise((resolve, reject) => {
        const collector = new ReactionCollector(client, messageId, options);
        collector.once('end', (collected, reason) => {
            if (options.max && collected.size < options.max) {
                reject(new Error(`Collector ended with reason: ${reason}`));
            }
            else {
                resolve(collected);
            }
        });
    });
}
exports.default = Collector;
