// @jubbio/core/dist/utils/Collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collection = void 0;
/**
 * A Map with additional utility methods
 */
class Collection extends Map {
    /**
     * Identical to Map.get()
     * Ensures the value exists
     */
    ensure(key, defaultValueGenerator) {
        if (this.has(key))
            return this.get(key);
        const defaultValue = defaultValueGenerator(key, this);
        this.set(key, defaultValue);
        return defaultValue;
    }
    /**
     * Checks if all items pass a test
     */
    every(fn) {
        for (const [key, val] of this) {
            if (!fn(val, key, this))
                return false;
        }
        return true;
    }
    /**
     * Checks if any item passes a test
     */
    some(fn) {
        for (const [key, val] of this) {
            if (fn(val, key, this))
                return true;
        }
        return false;
    }
    /**
     * Identical to Array.filter(), but returns a Collection
     */
    filter(fn) {
        const results = new Collection();
        for (const [key, val] of this) {
            if (fn(val, key, this))
                results.set(key, val);
        }
        return results;
    }
    /**
     * Partitions the collection into two collections
     */
    partition(fn) {
        const results = [new Collection(), new Collection()];
        for (const [key, val] of this) {
            if (fn(val, key, this)) {
                results[0].set(key, val);
            }
            else {
                results[1].set(key, val);
            }
        }
        return results;
    }
    /**
     * Maps each item to another value
     */
    map(fn) {
        const results = [];
        for (const [key, val] of this) {
            results.push(fn(val, key, this));
        }
        return results;
    }
    /**
     * Maps each item to another value into a Collection
     */
    mapValues(fn) {
        const results = new Collection();
        for (const [key, val] of this) {
            results.set(key, fn(val, key, this));
        }
        return results;
    }
    /**
     * Searches for a single item
     */
    find(fn) {
        for (const [key, val] of this) {
            if (fn(val, key, this))
                return val;
        }
        return undefined;
    }
    /**
     * Searches for the key of a single item
     */
    findKey(fn) {
        for (const [key, val] of this) {
            if (fn(val, key, this))
                return key;
        }
        return undefined;
    }
    /**
     * Removes items that satisfy the provided filter function
     */
    sweep(fn) {
        const previousSize = this.size;
        for (const [key, val] of this) {
            if (fn(val, key, this))
                this.delete(key);
        }
        return previousSize - this.size;
    }
    /**
     * Reduces the collection to a single value
     */
    reduce(fn, initialValue) {
        let accumulator = initialValue;
        for (const [key, val] of this) {
            accumulator = fn(accumulator, val, key, this);
        }
        return accumulator;
    }
    /**
     * Identical to Array.forEach()
     */
    each(fn) {
        for (const [key, val] of this) {
            fn(val, key, this);
        }
        return this;
    }
    first(amount) {
        if (amount === undefined)
            return this.values().next().value;
        if (amount < 0)
            return this.last(amount * -1);
        amount = Math.min(this.size, amount);
        const iter = this.values();
        return Array.from({ length: amount }, () => iter.next().value);
    }
    firstKey(amount) {
        if (amount === undefined)
            return this.keys().next().value;
        if (amount < 0)
            return this.lastKey(amount * -1);
        amount = Math.min(this.size, amount);
        const iter = this.keys();
        return Array.from({ length: amount }, () => iter.next().value);
    }
    last(amount) {
        const arr = [...this.values()];
        if (amount === undefined)
            return arr[arr.length - 1];
        if (amount < 0)
            return this.first(amount * -1);
        if (!amount)
            return [];
        return arr.slice(-amount);
    }
    lastKey(amount) {
        const arr = [...this.keys()];
        if (amount === undefined)
            return arr[arr.length - 1];
        if (amount < 0)
            return this.firstKey(amount * -1);
        if (!amount)
            return [];
        return arr.slice(-amount);
    }
    random(amount) {
        const arr = [...this.values()];
        if (amount === undefined)
            return arr[Math.floor(Math.random() * arr.length)];
        if (!arr.length || !amount)
            return [];
        return Array.from({ length: Math.min(amount, arr.length) }, () => arr.splice(Math.floor(Math.random() * arr.length), 1)[0]);
    }
    randomKey(amount) {
        const arr = [...this.keys()];
        if (amount === undefined)
            return arr[Math.floor(Math.random() * arr.length)];
        if (!arr.length || !amount)
            return [];
        return Array.from({ length: Math.min(amount, arr.length) }, () => arr.splice(Math.floor(Math.random() * arr.length), 1)[0]);
    }
    /**
     * Combines this collection with others
     */
    concat(...collections) {
        const newColl = this.clone();
        for (const coll of collections) {
            for (const [key, val] of coll)
                newColl.set(key, val);
        }
        return newColl;
    }
    /**
     * Checks if this collection shares identical items with another
     */
    equals(collection) {
        if (this === collection)
            return true;
        if (this.size !== collection.size)
            return false;
        for (const [key, value] of this) {
            if (!collection.has(key) || value !== collection.get(key)) {
                return false;
            }
        }
        return true;
    }
    /**
     * Creates an identical shallow copy of this collection
     */
    clone() {
        return new Collection(this);
    }
    /**
     * Sorts the collection and returns it
     */
    sort(compareFunction = Collection.defaultSort) {
        const entries = [...this.entries()];
        entries.sort((a, b) => compareFunction(a[1], b[1], a[0], b[0]));
        this.clear();
        for (const [k, v] of entries) {
            this.set(k, v);
        }
        return this;
    }
    /**
     * Sorts the collection by keys and returns it
     */
    sortByKey(compareFunction = Collection.defaultSort) {
        const entries = [...this.entries()];
        entries.sort((a, b) => compareFunction(a[0], b[0], a[1], b[1]));
        this.clear();
        for (const [k, v] of entries) {
            this.set(k, v);
        }
        return this;
    }
    /**
     * Returns an array of items
     */
    toJSON() {
        return [...this.values()];
    }
    /**
     * Default sort function
     */
    static defaultSort(firstValue, secondValue) {
        return Number(firstValue > secondValue) || Number(firstValue === secondValue) - 1;
    }
    /**
     * Creates a Collection from an array
     */
    static from(entries) {
        return new Collection(entries);
    }
}
exports.Collection = Collection;
