// @jubbio/core/dist/utils/Formatters.js
"use strict";
/**
 * Formatters for markdown and mentions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Formatters = exports.TimestampStyles = void 0;
exports.userMention = userMention;
exports.channelMention = channelMention;
exports.roleMention = roleMention;
exports.formatEmoji = formatEmoji;
exports.bold = bold;
exports.italic = italic;
exports.underline = underline;
exports.strikethrough = strikethrough;
exports.spoiler = spoiler;
exports.inlineCode = inlineCode;
exports.codeBlock = codeBlock;
exports.blockQuote = blockQuote;
exports.quote = quote;
exports.hyperlink = hyperlink;
exports.hideLinkEmbed = hideLinkEmbed;
exports.time = time;
exports.heading = heading;
exports.unorderedList = unorderedList;
exports.orderedList = orderedList;
/**
 * Formats a user mention
 * @param userId The user ID to mention
 */
function userMention(userId) {
    return `<@${userId}>`;
}
/**
 * Formats a channel mention
 * @param channelId The channel ID to mention
 */
function channelMention(channelId) {
    return `<#${channelId}>`;
}
/**
 * Formats a role mention
 * @param roleId The role ID to mention
 */
function roleMention(roleId) {
    return `<@&${roleId}>`;
}
/**
 * Formats a custom emoji
 * @param emojiId The emoji ID
 * @param name The emoji name
 * @param animated Whether the emoji is animated
 */
function formatEmoji(emojiId, name, animated = false) {
    return `<${animated ? 'a' : ''}:${name}:${emojiId}>`;
}
/**
 * Formats text as bold
 * @param text The text to format
 */
function bold(text) {
    return `**${text}**`;
}
/**
 * Formats text as italic
 * @param text The text to format
 */
function italic(text) {
    return `*${text}*`;
}
/**
 * Formats text as underline
 * @param text The text to format
 */
function underline(text) {
    return `__${text}__`;
}
/**
 * Formats text as strikethrough
 * @param text The text to format
 */
function strikethrough(text) {
    return `~~${text}~~`;
}
/**
 * Formats text as spoiler
 * @param text The text to format
 */
function spoiler(text) {
    return `||${text}||`;
}
/**
 * Formats text as inline code
 * @param text The text to format
 */
function inlineCode(text) {
    return `\`${text}\``;
}
/**
 * Formats text as a code block
 * @param text The text to format
 * @param language The language for syntax highlighting
 */
function codeBlock(text, language) {
    return `\`\`\`${language ?? ''}\n${text}\n\`\`\``;
}
/**
 * Formats text as a block quote
 * @param text The text to format
 */
function blockQuote(text) {
    return `>>> ${text}`;
}
/**
 * Formats text as a single-line quote
 * @param text The text to format
 */
function quote(text) {
    return `> ${text}`;
}
/**
 * Formats a URL as a hyperlink
 * @param text The text to display
 * @param url The URL to link to
 * @param title Optional title for the link
 */
function hyperlink(text, url, title) {
    return title ? `[${text}](${url} "${title}")` : `[${text}](${url})`;
}
/**
 * Formats a URL to hide the embed
 * @param url The URL to format
 */
function hideLinkEmbed(url) {
    return `<${url}>`;
}
/**
 * Time format styles
 */
var TimestampStyles;
(function (TimestampStyles) {
    /** Short time format (e.g., 16:20) */
    TimestampStyles["ShortTime"] = "t";
    /** Long time format (e.g., 16:20:30) */
    TimestampStyles["LongTime"] = "T";
    /** Short date format (e.g., 20/04/2021) */
    TimestampStyles["ShortDate"] = "d";
    /** Long date format (e.g., 20 April 2021) */
    TimestampStyles["LongDate"] = "D";
    /** Short date/time format (e.g., 20 April 2021 16:20) */
    TimestampStyles["ShortDateTime"] = "f";
    /** Long date/time format (e.g., Tuesday, 20 April 2021 16:20) */
    TimestampStyles["LongDateTime"] = "F";
    /** Relative time format (e.g., 2 months ago) */
    TimestampStyles["RelativeTime"] = "R";
})(TimestampStyles || (exports.TimestampStyles = TimestampStyles = {}));
/**
 * Formats a timestamp
 * @param timestamp The timestamp (Date, number in ms, or seconds)
 * @param style The style to use
 */
function time(timestamp, style) {
    const seconds = timestamp instanceof Date
        ? Math.floor(timestamp.getTime() / 1000)
        : typeof timestamp === 'number' && timestamp > 1e12
            ? Math.floor(timestamp / 1000)
            : timestamp;
    return style ? `<t:${seconds}:${style}>` : `<t:${seconds}>`;
}
/**
 * Formats a heading (H1)
 * @param text The text to format
 */
function heading(text, level = 1) {
    return `${'#'.repeat(level)} ${text}`;
}
/**
 * Formats an unordered list
 * @param items The items to list
 */
function unorderedList(items) {
    return items.map(item => `- ${item}`).join('\n');
}
/**
 * Formats an ordered list
 * @param items The items to list
 */
function orderedList(items) {
    return items.map((item, i) => `${i + 1}. ${item}`).join('\n');
}
// Export all formatters as a namespace too (DJS compatibility)
exports.Formatters = {
    userMention,
    channelMention,
    roleMention,
    formatEmoji,
    bold,
    italic,
    underline,
    strikethrough,
    spoiler,
    inlineCode,
    codeBlock,
    blockQuote,
    quote,
    hyperlink,
    hideLinkEmbed,
    time,
    heading,
    unorderedList,
    orderedList,
    TimestampStyles,
};
