// @jubbio/core/dist/utils/IntentsBitField.js
"use strict";
/**
 * Intents BitField for calculating gateway intents
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntentsBitField = exports.PrivilegedIntents = exports.IntentsAll = exports.GatewayIntentBits = void 0;
exports.calculateIntents = calculateIntents;
exports.resolveIntents = resolveIntents;
const BitField_1 = require("./BitField");
/**
 * Gateway Intent Bits
 */
exports.GatewayIntentBits = {
    Guilds: 1 << 0,
    GuildMembers: 1 << 1,
    GuildModeration: 1 << 2,
    GuildEmojisAndStickers: 1 << 3,
    GuildIntegrations: 1 << 4,
    GuildWebhooks: 1 << 5,
    GuildInvites: 1 << 6,
    GuildVoiceStates: 1 << 7,
    GuildPresences: 1 << 8,
    GuildMessages: 1 << 9,
    GuildMessageReactions: 1 << 10,
    GuildMessageTyping: 1 << 11,
    DirectMessages: 1 << 12,
    DirectMessageReactions: 1 << 13,
    DirectMessageTyping: 1 << 14,
    MessageContent: 1 << 15,
    GuildScheduledEvents: 1 << 16,
    AutoModerationConfiguration: 1 << 20,
    AutoModerationExecution: 1 << 21,
};
/**
 * All non-privileged intents
 */
exports.IntentsAll = Object.values(exports.GatewayIntentBits).reduce((all, i) => all | i, 0);
/**
 * Privileged intents that require approval
 */
exports.PrivilegedIntents = exports.GatewayIntentBits.GuildMembers |
    exports.GatewayIntentBits.GuildPresences |
    exports.GatewayIntentBits.MessageContent;
/**
 * Data structure for gateway intents
 */
class IntentsBitField extends BitField_1.BitField {
    static Flags = exports.GatewayIntentBits;
    static DefaultBit = 0;
    constructor(bits = 0) {
        super(bits);
    }
    /**
     * Check if any privileged intents are enabled
     */
    hasPrivileged() {
        return this.any(exports.PrivilegedIntents);
    }
    /**
     * Get all privileged intents that are enabled
     */
    getPrivileged() {
        const privileged = [];
        if (this.has(exports.GatewayIntentBits.GuildMembers))
            privileged.push('GuildMembers');
        if (this.has(exports.GatewayIntentBits.GuildPresences))
            privileged.push('GuildPresences');
        if (this.has(exports.GatewayIntentBits.MessageContent))
            privileged.push('MessageContent');
        return privileged;
    }
}
exports.IntentsBitField = IntentsBitField;
/**
 * Calculate intents from an array of intent names or values
 */
function calculateIntents(intents) {
    return intents.reduce((acc, intent) => {
        if (typeof intent === 'number')
            return acc | intent;
        if (intent in exports.GatewayIntentBits)
            return acc | exports.GatewayIntentBits[intent];
        return acc;
    }, 0);
}
/**
 * Get all intent names from a bitfield value
 */
function resolveIntents(bits) {
    const names = [];
    for (const [name, value] of Object.entries(exports.GatewayIntentBits)) {
        if ((bits & value) === value) {
            names.push(name);
        }
    }
    return names;
}
exports.default = IntentsBitField;
