// @jubbio/core/dist/utils/Partials.js
"use strict";
/**
 * Partials - Handle uncached/partial data structures
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Partials = void 0;
exports.isPartial = isPartial;
exports.createPartialUser = createPartialUser;
exports.createPartialChannel = createPartialChannel;
exports.createPartialMessage = createPartialMessage;
exports.createPartialGuildMember = createPartialGuildMember;
exports.createPartialReaction = createPartialReaction;
exports.makePartialAware = makePartialAware;
exports.hasPartial = hasPartial;
/**
 * Partial types that can be enabled
 */
var Partials;
(function (Partials) {
    Partials[Partials["User"] = 0] = "User";
    Partials[Partials["Channel"] = 1] = "Channel";
    Partials[Partials["GuildMember"] = 2] = "GuildMember";
    Partials[Partials["Message"] = 3] = "Message";
    Partials[Partials["Reaction"] = 4] = "Reaction";
    Partials[Partials["GuildScheduledEvent"] = 5] = "GuildScheduledEvent";
})(Partials || (exports.Partials = Partials = {}));
/**
 * Check if a structure is partial (missing data)
 */
function isPartial(obj) {
    return obj?.partial === true;
}
/**
 * Create a partial user structure
 */
function createPartialUser(id) {
    return {
        id,
        partial: true,
        username: null,
        discriminator: null,
        avatar: null,
        bot: null,
        async fetch() {
            throw new Error('Cannot fetch partial user without client context');
        },
        toString() {
            return `<@${id}>`;
        },
    };
}
/**
 * Create a partial channel structure
 */
function createPartialChannel(id) {
    return {
        id,
        partial: true,
        type: null,
        name: null,
        async fetch() {
            throw new Error('Cannot fetch partial channel without client context');
        },
        toString() {
            return `<#${id}>`;
        },
    };
}
/**
 * Create a partial message structure
 */
function createPartialMessage(id, channelId) {
    return {
        id,
        channelId,
        partial: true,
        content: null,
        author: null,
        embeds: null,
        attachments: null,
        async fetch() {
            throw new Error('Cannot fetch partial message without client context');
        },
        toString() {
            return `Message(${id})`;
        },
    };
}
/**
 * Create a partial guild member structure
 */
function createPartialGuildMember(userId, guildId) {
    return {
        id: userId,
        guildId,
        partial: true,
        user: null,
        nick: null,
        roles: null,
        joinedAt: null,
        async fetch() {
            throw new Error('Cannot fetch partial member without client context');
        },
        toString() {
            return `<@${userId}>`;
        },
    };
}
/**
 * Create a partial reaction structure
 */
function createPartialReaction(messageId, emoji) {
    return {
        messageId,
        emoji,
        partial: true,
        count: null,
        me: null,
        async fetch() {
            throw new Error('Cannot fetch partial reaction without client context');
        },
        toString() {
            return emoji;
        },
    };
}
/**
 * Make a structure partial-aware with fetch capability
 */
function makePartialAware(structure, client, fetchFn) {
    return {
        ...structure,
        partial: false,
        async fetch() {
            const fetched = await fetchFn(structure.id);
            Object.assign(this, fetched, { partial: false });
            return this;
        },
    };
}
/**
 * Check if partials are enabled for a type
 */
function hasPartial(client, partial) {
    return client.options?.partials?.includes(partial) ?? false;
}
exports.default = Partials;
