// @jubbio/core/dist/utils/PermissionsBitField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PermissionsBitField = void 0;
const enums_1 = require("../enums");
/**
 * Bit field for permissions
 * API compatible with Discord.js PermissionsBitField
 */
class PermissionsBitField {
    /** The raw bits */
    bitfield;
    /** All permission flags */
    static Flags = enums_1.PermissionFlagsBits;
    /** All permissions combined */
    static All = Object.values(enums_1.PermissionFlagsBits).reduce((acc, val) => acc | val, 0n);
    /** Default permissions */
    static Default = BigInt(0);
    constructor(bits = 0n) {
        this.bitfield = PermissionsBitField.resolve(bits);
    }
    /**
     * Check if this bitfield has a permission
     */
    has(permission, checkAdmin = true) {
        // Admin has all permissions
        if (checkAdmin && this.bitfield & enums_1.PermissionFlagsBits.Administrator) {
            return true;
        }
        const bit = PermissionsBitField.resolve(permission);
        return (this.bitfield & bit) === bit;
    }
    /**
     * Check if this bitfield has any of the permissions
     */
    any(permissions, checkAdmin = true) {
        // Admin has all permissions
        if (checkAdmin && this.bitfield & enums_1.PermissionFlagsBits.Administrator) {
            return true;
        }
        const bit = PermissionsBitField.resolve(permissions);
        return (this.bitfield & bit) !== 0n;
    }
    /**
     * Check if this bitfield is missing any permissions
     */
    missing(permissions, checkAdmin = true) {
        const missing = [];
        for (const [name, bit] of Object.entries(enums_1.PermissionFlagsBits)) {
            const resolved = PermissionsBitField.resolve(permissions);
            if ((resolved & bit) && !this.has(bit, checkAdmin)) {
                missing.push(name);
            }
        }
        return missing;
    }
    /**
     * Add permissions to this bitfield
     */
    add(...permissions) {
        for (const permission of permissions) {
            this.bitfield |= PermissionsBitField.resolve(permission);
        }
        return this;
    }
    /**
     * Remove permissions from this bitfield
     */
    remove(...permissions) {
        for (const permission of permissions) {
            this.bitfield &= ~PermissionsBitField.resolve(permission);
        }
        return this;
    }
    /**
     * Serialize this bitfield to an array of permission names
     */
    toArray() {
        const result = [];
        for (const [name, bit] of Object.entries(enums_1.PermissionFlagsBits)) {
            if (this.bitfield & bit) {
                result.push(name);
            }
        }
        return result;
    }
    /**
     * Serialize this bitfield to a JSON-compatible value
     */
    toJSON() {
        return this.bitfield.toString();
    }
    /**
     * Get the string representation
     */
    toString() {
        return this.bitfield.toString();
    }
    /**
     * Freeze this bitfield
     */
    freeze() {
        return Object.freeze(this);
    }
    /**
     * Check equality with another bitfield
     */
    equals(other) {
        return this.bitfield === PermissionsBitField.resolve(other);
    }
    /**
     * Create a new bitfield with the same bits
     */
    clone() {
        return new PermissionsBitField(this.bitfield);
    }
    /**
     * Resolve a permission to a bigint
     */
    static resolve(permission) {
        if (typeof permission === 'bigint') {
            return permission;
        }
        if (permission instanceof PermissionsBitField) {
            return permission.bitfield;
        }
        if (typeof permission === 'string') {
            const bit = enums_1.PermissionFlagsBits[permission];
            if (bit === undefined) {
                throw new Error(`Unknown permission: ${permission}`);
            }
            return bit;
        }
        if (Array.isArray(permission)) {
            let result = 0n;
            for (const p of permission) {
                result |= PermissionsBitField.resolve(p);
            }
            return result;
        }
        throw new Error(`Invalid permission: ${permission}`);
    }
}
exports.PermissionsBitField = PermissionsBitField;
exports.default = PermissionsBitField;
