// @jubbio/core/dist/utils/SnowflakeUtil.js
"use strict";
/**
 * Snowflake utilities
 * API compatible with Discord.js SnowflakeUtil
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnowflakeUtil = void 0;
// Jubbio epoch (same as Discord: 2015-01-01T00:00:00.000Z)
const EPOCH = 1420070400000n;
/**
 * A container for useful snowflake-related methods
 */
class SnowflakeUtil {
    /**
     * Jubbio's epoch value
     */
    static EPOCH = EPOCH;
    /**
     * Generates a snowflake ID
     * @param timestamp Timestamp or date to generate from
     */
    static generate(timestamp = Date.now()) {
        const time = timestamp instanceof Date ? timestamp.getTime() : timestamp;
        return ((BigInt(time) - EPOCH) << 22n).toString();
    }
    /**
     * Deconstructs a snowflake ID
     * @param snowflake Snowflake to deconstruct
     */
    static deconstruct(snowflake) {
        const bigIntSnowflake = BigInt(snowflake);
        return {
            timestamp: Number((bigIntSnowflake >> 22n) + EPOCH),
            get date() {
                return new Date(this.timestamp);
            },
            workerId: Number((bigIntSnowflake & 0x3e0000n) >> 17n),
            processId: Number((bigIntSnowflake & 0x1f000n) >> 12n),
            increment: Number(bigIntSnowflake & 0xfffn),
            binary: bigIntSnowflake.toString(2).padStart(64, '0'),
        };
    }
    /**
     * Retrieves the timestamp from a snowflake
     * @param snowflake Snowflake to get the timestamp from
     */
    static timestampFrom(snowflake) {
        return Number((BigInt(snowflake) >> 22n) + EPOCH);
    }
    /**
     * Retrieves the date from a snowflake
     * @param snowflake Snowflake to get the date from
     */
    static dateFrom(snowflake) {
        return new Date(SnowflakeUtil.timestampFrom(snowflake));
    }
    /**
     * Compares two snowflakes
     * @param a First snowflake
     * @param b Second snowflake
     * @returns -1 if a < b, 0 if a === b, 1 if a > b
     */
    static compare(a, b) {
        const bigA = BigInt(a);
        const bigB = BigInt(b);
        if (bigA < bigB)
            return -1;
        if (bigA > bigB)
            return 1;
        return 0;
    }
    /**
     * Checks if a value is a valid snowflake
     * @param value Value to check
     */
    static isValid(value) {
        if (typeof value !== 'string')
            return false;
        if (!/^\d{17,20}$/.test(value))
            return false;
        try {
            const timestamp = SnowflakeUtil.timestampFrom(value);
            return timestamp > Number(EPOCH) && timestamp < Date.now() + 1000 * 60 * 60 * 24 * 365; // Within reasonable range
        }
        catch {
            return false;
        }
    }
}
exports.SnowflakeUtil = SnowflakeUtil;
// Export as default too for convenience
exports.default = SnowflakeUtil;
