// @jubbio/core/dist/utils/Sweepers.js
"use strict";
/**
 * Sweepers - Automatic cache cleanup utilities
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultSweeperOptions = exports.SweeperManager = exports.Sweepers = void 0;
/**
 * Sweeper filters - predefined filter functions
 */
exports.Sweepers = {
    /**
     * Filter that sweeps items older than a certain lifetime
     */
    filterByLifetime(options = {}) {
        const lifetime = options.lifetime ?? 14400; // 4 hours default
        const getTimestamp = options.getComparisonTimestamp ??
            ((v) => v.createdTimestamp ?? v.createdAt?.getTime() ?? 0);
        const exclude = options.excludeFromSweep ?? (() => false);
        return () => {
            const now = Date.now();
            const cutoff = now - (lifetime * 1000);
            return (value) => {
                if (exclude(value))
                    return false;
                const timestamp = getTimestamp(value);
                return timestamp < cutoff;
            };
        };
    },
    /**
     * Filter that sweeps expired invites
     */
    expiredInviteSweepFilter() {
        return () => {
            const now = Date.now();
            return (invite) => {
                if (!invite.expiresAt && !invite.expiresTimestamp)
                    return false;
                const expiresAt = invite.expiresAt?.getTime() ?? invite.expiresTimestamp ?? Infinity;
                return expiresAt < now;
            };
        };
    },
    /**
     * Filter that sweeps outdated presences
     */
    outdatedPresenceSweepFilter(lifetime = 21600) {
        return () => {
            const now = Date.now();
            const cutoff = now - (lifetime * 1000);
            return (presence) => {
                const lastUpdate = presence.lastModified ?? presence.updatedAt?.getTime() ?? 0;
                return lastUpdate < cutoff;
            };
        };
    },
    /**
     * Filter that sweeps all items (use with caution)
     */
    sweepAll() {
        return () => () => true;
    },
    /**
     * Filter that sweeps nothing
     */
    sweepNone() {
        return () => () => false;
    },
};
/**
 * Sweeper manager class
 */
class SweeperManager {
    client;
    intervals = new Map();
    options;
    constructor(client, options = {}) {
        this.client = client;
        this.options = options;
    }
    /**
     * Start all configured sweepers
     */
    start() {
        for (const [key, config] of Object.entries(this.options)) {
            if (config && config.interval > 0) {
                this.startSweeper(key, config);
            }
        }
    }
    /**
     * Start a specific sweeper
     */
    startSweeper(name, config) {
        // Clear existing interval if any
        this.stopSweeper(name);
        const interval = setInterval(() => {
            this.sweep(name, config.filter);
        }, config.interval * 1000);
        this.intervals.set(name, interval);
    }
    /**
     * Stop a specific sweeper
     */
    stopSweeper(name) {
        const interval = this.intervals.get(name);
        if (interval) {
            clearInterval(interval);
            this.intervals.delete(name);
        }
    }
    /**
     * Stop all sweepers
     */
    stop() {
        for (const name of this.intervals.keys()) {
            this.stopSweeper(name);
        }
    }
    /**
     * Manually trigger a sweep
     */
    sweep(name, filter) {
        const cache = this.getCache(name);
        if (!cache)
            return 0;
        const filterFn = filter ?? this.options[name]?.filter;
        if (!filterFn)
            return 0;
        return cache.sweep(filterFn());
    }
    /**
     * Get the cache for a sweeper type
     */
    getCache(name) {
        switch (name) {
            case 'users':
                return this.client.users?.cache ?? null;
            case 'guildMembers':
                // Sweep across all guilds
                let memberCount = 0;
                this.client.guilds?.cache.forEach((guild) => {
                    if (guild.members?.cache) {
                        const filter = this.options.guildMembers?.filter;
                        if (filter)
                            memberCount += guild.members.cache.sweep(filter());
                    }
                });
                return null; // Already handled
            case 'messages':
                // Sweep across all channels
                this.client.channels?.cache.forEach((channel) => {
                    if (channel.messages?.cache) {
                        const filter = this.options.messages?.filter;
                        if (filter)
                            channel.messages.cache.sweep(filter());
                    }
                });
                return null;
            case 'presences':
                // Sweep across all guilds
                this.client.guilds?.cache.forEach((guild) => {
                    if (guild.presences?.cache) {
                        const filter = this.options.presences?.filter;
                        if (filter)
                            guild.presences.cache.sweep(filter());
                    }
                });
                return null;
            case 'voiceStates':
                this.client.guilds?.cache.forEach((guild) => {
                    if (guild.voiceStates?.cache) {
                        const filter = this.options.voiceStates?.filter;
                        if (filter)
                            guild.voiceStates.cache.sweep(filter());
                    }
                });
                return null;
            case 'reactions':
                // Sweep reactions from all messages
                this.client.channels?.cache.forEach((channel) => {
                    channel.messages?.cache.forEach((message) => {
                        if (message.reactions?.cache) {
                            const filter = this.options.reactions?.filter;
                            if (filter)
                                message.reactions.cache.sweep(filter());
                        }
                    });
                });
                return null;
            case 'emojis':
                return this.client.emojis?.cache ?? null;
            case 'stickers':
                return this.client.stickers?.cache ?? null;
            case 'invites':
                // Sweep across all guilds
                this.client.guilds?.cache.forEach((guild) => {
                    if (guild.invites?.cache) {
                        const filter = this.options.invites?.filter;
                        if (filter)
                            guild.invites.cache.sweep(filter());
                    }
                });
                return null;
            case 'bans':
                this.client.guilds?.cache.forEach((guild) => {
                    if (guild.bans?.cache) {
                        const filter = this.options.bans?.filter;
                        if (filter)
                            guild.bans.cache.sweep(filter());
                    }
                });
                return null;
            default:
                return null;
        }
    }
    /**
     * Get sweeper statistics
     */
    getStats() {
        const stats = {};
        for (const [name, config] of Object.entries(this.options)) {
            if (config) {
                stats[name] = {
                    interval: config.interval,
                    running: this.intervals.has(name),
                };
            }
        }
        return stats;
    }
}
exports.SweeperManager = SweeperManager;
/**
 * Default sweeper options for common use cases
 */
exports.DefaultSweeperOptions = {
    messages: {
        interval: 3600, // 1 hour
        filter: exports.Sweepers.filterByLifetime({ lifetime: 1800 }), // 30 minutes
    },
    invites: {
        interval: 3600,
        filter: exports.Sweepers.expiredInviteSweepFilter(),
    },
};
exports.default = SweeperManager;
