// @jubbio/core/dist/builders/ButtonBuilder.js
"use strict";
/**
 * ButtonBuilder for creating interactive buttons
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ButtonBuilder = exports.ButtonStyle = void 0;
/**
 * Button styles
 */
var ButtonStyle;
(function (ButtonStyle) {
    ButtonStyle[ButtonStyle["Primary"] = 1] = "Primary";
    ButtonStyle[ButtonStyle["Secondary"] = 2] = "Secondary";
    ButtonStyle[ButtonStyle["Success"] = 3] = "Success";
    ButtonStyle[ButtonStyle["Danger"] = 4] = "Danger";
    ButtonStyle[ButtonStyle["Link"] = 5] = "Link";
})(ButtonStyle || (exports.ButtonStyle = ButtonStyle = {}));
/**
 * A builder for creating buttons
 */
class ButtonBuilder {
    data;
    constructor(data = {}) {
        this.data = { type: 2, ...data };
    }
    /**
     * Sets the custom ID of this button
     * @param customId The custom ID
     */
    setCustomId(customId) {
        this.data.custom_id = customId;
        return this;
    }
    /**
     * Sets the label of this button
     * @param label The label
     */
    setLabel(label) {
        this.data.label = label;
        return this;
    }
    /**
     * Sets the style of this button
     * @param style The style
     */
    setStyle(style) {
        this.data.style = style;
        return this;
    }
    /**
     * Sets the emoji of this button
     * @param emoji The emoji
     */
    setEmoji(emoji) {
        if (typeof emoji === 'string') {
            // Unicode emoji
            this.data.emoji = { name: emoji };
        }
        else {
            this.data.emoji = emoji;
        }
        return this;
    }
    /**
     * Sets the URL of this button (only for Link style)
     * @param url The URL
     */
    setURL(url) {
        this.data.url = url;
        return this;
    }
    /**
     * Sets whether this button is disabled
     * @param disabled Whether the button is disabled
     */
    setDisabled(disabled = true) {
        this.data.disabled = disabled;
        return this;
    }
    /**
     * Returns the JSON representation of this button
     */
    toJSON() {
        return { ...this.data };
    }
    /**
     * Creates a new button builder from existing data
     * @param other The button data to copy
     */
    static from(other) {
        return new ButtonBuilder(other instanceof ButtonBuilder ? other.data : other);
    }
}
exports.ButtonBuilder = ButtonBuilder;
exports.default = ButtonBuilder;
