// @jubbio/core/dist/builders/ModalBuilder.js
"use strict";
/**
 * ModalBuilder for creating modals
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModalBuilder = exports.TextInputBuilder = exports.TextInputStyle = void 0;
/**
 * Text input styles
 */
var TextInputStyle;
(function (TextInputStyle) {
    TextInputStyle[TextInputStyle["Short"] = 1] = "Short";
    TextInputStyle[TextInputStyle["Paragraph"] = 2] = "Paragraph";
})(TextInputStyle || (exports.TextInputStyle = TextInputStyle = {}));
/**
 * A builder for creating text inputs
 */
class TextInputBuilder {
    data;
    constructor(data = {}) {
        this.data = { type: 4, ...data };
    }
    /**
     * Sets the custom ID of this text input
     * @param customId The custom ID
     */
    setCustomId(customId) {
        this.data.custom_id = customId;
        return this;
    }
    /**
     * Sets the label of this text input
     * @param label The label
     */
    setLabel(label) {
        this.data.label = label;
        return this;
    }
    /**
     * Sets the style of this text input
     * @param style The style
     */
    setStyle(style) {
        this.data.style = style;
        return this;
    }
    /**
     * Sets the minimum length of this text input
     * @param minLength The minimum length
     */
    setMinLength(minLength) {
        this.data.min_length = minLength;
        return this;
    }
    /**
     * Sets the maximum length of this text input
     * @param maxLength The maximum length
     */
    setMaxLength(maxLength) {
        this.data.max_length = maxLength;
        return this;
    }
    /**
     * Sets whether this text input is required
     * @param required Whether the text input is required
     */
    setRequired(required = true) {
        this.data.required = required;
        return this;
    }
    /**
     * Sets the value of this text input
     * @param value The value
     */
    setValue(value) {
        this.data.value = value;
        return this;
    }
    /**
     * Sets the placeholder of this text input
     * @param placeholder The placeholder
     */
    setPlaceholder(placeholder) {
        this.data.placeholder = placeholder;
        return this;
    }
    /**
     * Returns the JSON representation of this text input
     */
    toJSON() {
        return { ...this.data };
    }
}
exports.TextInputBuilder = TextInputBuilder;
/**
 * A builder for creating modals
 */
class ModalBuilder {
    data;
    constructor(data = {}) {
        this.data = { ...data };
        if (!this.data.components)
            this.data.components = [];
    }
    /**
     * Sets the custom ID of this modal
     * @param customId The custom ID
     */
    setCustomId(customId) {
        this.data.custom_id = customId;
        return this;
    }
    /**
     * Sets the title of this modal
     * @param title The title
     */
    setTitle(title) {
        this.data.title = title;
        return this;
    }
    /**
     * Adds components (action rows with text inputs) to this modal
     * @param components The components to add
     */
    addComponents(...components) {
        if (!this.data.components)
            this.data.components = [];
        for (const component of components) {
            if ('toJSON' in component && typeof component.toJSON === 'function') {
                this.data.components.push(component.toJSON());
            }
            else {
                this.data.components.push(component);
            }
        }
        return this;
    }
    /**
     * Sets the components of this modal
     * @param components The components to set
     */
    setComponents(...components) {
        this.data.components = [];
        return this.addComponents(...components);
    }
    /**
     * Returns the JSON representation of this modal
     */
    toJSON() {
        return { ...this.data };
    }
    /**
     * Creates a new modal builder from existing data
     * @param other The modal data to copy
     */
    static from(other) {
        return new ModalBuilder(other instanceof ModalBuilder ? other.data : other);
    }
}
exports.ModalBuilder = ModalBuilder;
exports.default = ModalBuilder;
