// @jubbio/core/dist/builders/SlashCommandBuilder.js
"use strict";
/**
 * SlashCommandBuilder for creating slash commands
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlashCommandBuilder = exports.SlashCommandAttachmentOption = exports.SlashCommandMentionableOption = exports.SlashCommandRoleOption = exports.SlashCommandChannelOption = exports.SlashCommandUserOption = exports.SlashCommandBooleanOption = exports.SlashCommandNumberOption = exports.SlashCommandIntegerOption = exports.SlashCommandStringOption = exports.ApplicationCommandOptionType = void 0;
/**
 * Application command option types
 */
var ApplicationCommandOptionType;
(function (ApplicationCommandOptionType) {
    ApplicationCommandOptionType[ApplicationCommandOptionType["Subcommand"] = 1] = "Subcommand";
    ApplicationCommandOptionType[ApplicationCommandOptionType["SubcommandGroup"] = 2] = "SubcommandGroup";
    ApplicationCommandOptionType[ApplicationCommandOptionType["String"] = 3] = "String";
    ApplicationCommandOptionType[ApplicationCommandOptionType["Integer"] = 4] = "Integer";
    ApplicationCommandOptionType[ApplicationCommandOptionType["Boolean"] = 5] = "Boolean";
    ApplicationCommandOptionType[ApplicationCommandOptionType["User"] = 6] = "User";
    ApplicationCommandOptionType[ApplicationCommandOptionType["Channel"] = 7] = "Channel";
    ApplicationCommandOptionType[ApplicationCommandOptionType["Role"] = 8] = "Role";
    ApplicationCommandOptionType[ApplicationCommandOptionType["Mentionable"] = 9] = "Mentionable";
    ApplicationCommandOptionType[ApplicationCommandOptionType["Number"] = 10] = "Number";
    ApplicationCommandOptionType[ApplicationCommandOptionType["Attachment"] = 11] = "Attachment";
})(ApplicationCommandOptionType || (exports.ApplicationCommandOptionType = ApplicationCommandOptionType = {}));
/**
 * A builder for creating slash command options
 */
class SlashCommandOptionBuilder {
    data;
    constructor(type) {
        this.data = { type };
    }
    setName(name) {
        this.data.name = name;
        return this;
    }
    setDescription(description) {
        this.data.description = description;
        return this;
    }
    setRequired(required = true) {
        this.data.required = required;
        return this;
    }
    toJSON() {
        return { ...this.data };
    }
}
/**
 * String option builder
 */
class SlashCommandStringOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.String);
    }
    addChoices(...choices) {
        if (!this.data.choices)
            this.data.choices = [];
        this.data.choices.push(...choices);
        return this;
    }
    setChoices(...choices) {
        this.data.choices = choices;
        return this;
    }
    setMinLength(minLength) {
        this.data.min_length = minLength;
        return this;
    }
    setMaxLength(maxLength) {
        this.data.max_length = maxLength;
        return this;
    }
    setAutocomplete(autocomplete = true) {
        this.data.autocomplete = autocomplete;
        return this;
    }
}
exports.SlashCommandStringOption = SlashCommandStringOption;
/**
 * Integer option builder
 */
class SlashCommandIntegerOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.Integer);
    }
    addChoices(...choices) {
        if (!this.data.choices)
            this.data.choices = [];
        this.data.choices.push(...choices);
        return this;
    }
    setMinValue(minValue) {
        this.data.min_value = minValue;
        return this;
    }
    setMaxValue(maxValue) {
        this.data.max_value = maxValue;
        return this;
    }
    setAutocomplete(autocomplete = true) {
        this.data.autocomplete = autocomplete;
        return this;
    }
}
exports.SlashCommandIntegerOption = SlashCommandIntegerOption;
/**
 * Number option builder
 */
class SlashCommandNumberOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.Number);
    }
    addChoices(...choices) {
        if (!this.data.choices)
            this.data.choices = [];
        this.data.choices.push(...choices);
        return this;
    }
    setMinValue(minValue) {
        this.data.min_value = minValue;
        return this;
    }
    setMaxValue(maxValue) {
        this.data.max_value = maxValue;
        return this;
    }
    setAutocomplete(autocomplete = true) {
        this.data.autocomplete = autocomplete;
        return this;
    }
}
exports.SlashCommandNumberOption = SlashCommandNumberOption;
/**
 * Boolean option builder
 */
class SlashCommandBooleanOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.Boolean);
    }
}
exports.SlashCommandBooleanOption = SlashCommandBooleanOption;
/**
 * User option builder
 */
class SlashCommandUserOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.User);
    }
}
exports.SlashCommandUserOption = SlashCommandUserOption;
/**
 * Channel option builder
 */
class SlashCommandChannelOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.Channel);
    }
    addChannelTypes(...types) {
        if (!this.data.channel_types)
            this.data.channel_types = [];
        this.data.channel_types.push(...types);
        return this;
    }
}
exports.SlashCommandChannelOption = SlashCommandChannelOption;
/**
 * Role option builder
 */
class SlashCommandRoleOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.Role);
    }
}
exports.SlashCommandRoleOption = SlashCommandRoleOption;
/**
 * Mentionable option builder
 */
class SlashCommandMentionableOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.Mentionable);
    }
}
exports.SlashCommandMentionableOption = SlashCommandMentionableOption;
/**
 * Attachment option builder
 */
class SlashCommandAttachmentOption extends SlashCommandOptionBuilder {
    constructor() {
        super(ApplicationCommandOptionType.Attachment);
    }
}
exports.SlashCommandAttachmentOption = SlashCommandAttachmentOption;
/**
 * A builder for creating slash commands
 */
class SlashCommandBuilder {
    data;
    constructor() {
        this.data = { options: [] };
    }
    /**
     * Sets the name of this command
     * @param name The name
     */
    setName(name) {
        this.data.name = name;
        return this;
    }
    /**
     * Sets the description of this command
     * @param description The description
     */
    setDescription(description) {
        this.data.description = description;
        return this;
    }
    /**
     * Sets the default member permissions required to use this command
     * @param permissions The permissions
     */
    setDefaultMemberPermissions(permissions) {
        this.data.default_member_permissions = permissions === null
            ? undefined
            : String(permissions);
        return this;
    }
    /**
     * Sets whether this command is available in DMs
     * @param enabled Whether the command is available in DMs
     */
    setDMPermission(enabled) {
        this.data.dm_permission = enabled;
        return this;
    }
    /**
     * Sets whether this command is NSFW
     * @param nsfw Whether the command is NSFW
     */
    setNSFW(nsfw = true) {
        this.data.nsfw = nsfw;
        return this;
    }
    /**
     * Adds a string option
     */
    addStringOption(fn) {
        const option = fn(new SlashCommandStringOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds an integer option
     */
    addIntegerOption(fn) {
        const option = fn(new SlashCommandIntegerOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds a number option
     */
    addNumberOption(fn) {
        const option = fn(new SlashCommandNumberOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds a boolean option
     */
    addBooleanOption(fn) {
        const option = fn(new SlashCommandBooleanOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds a user option
     */
    addUserOption(fn) {
        const option = fn(new SlashCommandUserOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds a channel option
     */
    addChannelOption(fn) {
        const option = fn(new SlashCommandChannelOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds a role option
     */
    addRoleOption(fn) {
        const option = fn(new SlashCommandRoleOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds a mentionable option
     */
    addMentionableOption(fn) {
        const option = fn(new SlashCommandMentionableOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Adds an attachment option
     */
    addAttachmentOption(fn) {
        const option = fn(new SlashCommandAttachmentOption());
        this.data.options.push(option.toJSON());
        return this;
    }
    /**
     * Returns the JSON representation of this command
     */
    toJSON() {
        return { ...this.data };
    }
}
exports.SlashCommandBuilder = SlashCommandBuilder;
exports.default = SlashCommandBuilder;
