// @jubbio/core/dist/managers/BaseManager.js
"use strict";
/**
 * Base manager class for caching and managing structures
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataManager = exports.CachedManager = exports.BaseManager = void 0;
const Collection_1 = require("../utils/Collection");
/**
 * Base manager for caching structures
 */
class BaseManager {
    /** The client that instantiated this manager */
    client;
    /** The cache of items */
    cache;
    /** The class to instantiate for items */
    holds;
    constructor(client, holds, iterable) {
        this.client = client;
        this.holds = holds;
        this.cache = new Collection_1.Collection();
        if (iterable) {
            for (const item of iterable) {
                this._add(item);
            }
        }
    }
    /**
     * Resolve an item from the cache or ID
     */
    resolve(idOrInstance) {
        if (idOrInstance instanceof this.holds)
            return idOrInstance;
        if (typeof idOrInstance === 'string')
            return this.cache.get(idOrInstance) ?? null;
        return null;
    }
    /**
     * Resolve an ID from an item or ID
     */
    resolveId(idOrInstance) {
        if (idOrInstance instanceof this.holds)
            return idOrInstance.id;
        if (typeof idOrInstance === 'string')
            return idOrInstance;
        if (typeof idOrInstance === 'object' && idOrInstance !== null && 'id' in idOrInstance) {
            return idOrInstance.id;
        }
        return null;
    }
    /**
     * Get the cache as a JSON array
     */
    valueOf() {
        return [...this.cache.values()];
    }
}
exports.BaseManager = BaseManager;
/**
 * Caching manager that fetches data from the API
 */
class CachedManager extends BaseManager {
    /**
     * Fetch an item, using cache if available (async version)
     */
    async resolveAsync(idOrInstance, options) {
        const existing = super.resolve(idOrInstance);
        if (existing && !options?.force)
            return existing;
        const id = this.resolveId(idOrInstance);
        if (!id)
            return null;
        try {
            return await this.fetch(id, options);
        }
        catch {
            return null;
        }
    }
}
exports.CachedManager = CachedManager;
/**
 * Data manager that doesn't cache
 */
class DataManager extends BaseManager {
    /**
     * Add an item to the cache
     */
    _add(data, cache = true, options) {
        const existing = this.cache.get(options?.id ?? data.id);
        if (existing) {
            if (cache) {
                existing._patch?.(data);
            }
            return existing;
        }
        const entry = new this.holds(this.client, data, ...(options?.extras ?? []));
        if (cache) {
            this.cache.set(options?.id ?? data.id, entry);
        }
        return entry;
    }
}
exports.DataManager = DataManager;
exports.default = BaseManager;
