// @jubbio/core/dist/managers/ChannelManager.js
"use strict";
/**
 * Manager for channels with caching and lazy loading
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelManager = exports.GuildChannelManager = exports.ChannelType = void 0;
const BaseManager_1 = require("./BaseManager");
const Collection_1 = require("../utils/Collection");
/** Channel types */
var ChannelType;
(function (ChannelType) {
    ChannelType[ChannelType["GuildText"] = 0] = "GuildText";
    ChannelType[ChannelType["DM"] = 1] = "DM";
    ChannelType[ChannelType["GuildVoice"] = 2] = "GuildVoice";
    ChannelType[ChannelType["GroupDM"] = 3] = "GroupDM";
    ChannelType[ChannelType["GuildCategory"] = 4] = "GuildCategory";
    ChannelType[ChannelType["GuildAnnouncement"] = 5] = "GuildAnnouncement";
    ChannelType[ChannelType["GuildStageVoice"] = 13] = "GuildStageVoice";
    ChannelType[ChannelType["GuildDirectory"] = 14] = "GuildDirectory";
    ChannelType[ChannelType["GuildForum"] = 15] = "GuildForum";
})(ChannelType || (exports.ChannelType = ChannelType = {}));
/**
 * Manages channels for a guild
 */
class GuildChannelManager extends BaseManager_1.CachedManager {
    /** The guild this manager belongs to */
    guild;
    constructor(guild, iterable) {
        super(guild.client, Object, iterable);
        this.guild = guild;
    }
    /**
     * Add a channel to the cache
     */
    _add(data, cache = true) {
        const id = data.id;
        const existing = this.cache.get(id);
        if (existing) {
            if (cache)
                Object.assign(existing, data);
            return existing;
        }
        const channel = {
            id,
            guildId: this.guild.id,
            name: data.name,
            type: data.type,
            position: data.position ?? 0,
            parentId: data.parent_id ?? null,
            permissionOverwrites: data.permission_overwrites ?? [],
            topic: data.topic ?? null,
            nsfw: data.nsfw ?? false,
            rateLimitPerUser: data.rate_limit_per_user ?? 0,
            bitrate: data.bitrate,
            userLimit: data.user_limit,
            get isText() { return [ChannelType.GuildText, ChannelType.GuildAnnouncement].includes(this.type); },
            get isVoice() { return [ChannelType.GuildVoice, ChannelType.GuildStageVoice].includes(this.type); },
            get isCategory() { return this.type === ChannelType.GuildCategory; },
            toString() { return `<#${this.id}>`; },
            async send(content) {
                return this.guild.client.rest.request('POST', `/channels/${this.id}/messages`, typeof content === 'string' ? { content } : content);
            },
            async delete(reason) {
                return this.guild.client.rest.request('DELETE', `/channels/${this.id}`, reason ? { reason } : undefined);
            },
            async edit(data) {
                return this.guild.client.rest.request('PATCH', `/channels/${this.id}`, data);
            },
        };
        if (cache)
            this.cache.set(id, channel);
        return channel;
    }
    /**
     * Fetch a channel from the API
     */
    async fetch(id, options) {
        if (!options?.force) {
            const existing = this.cache.get(id);
            if (existing)
                return existing;
        }
        const data = await this.client.rest.request('GET', `/channels/${id}`);
        return this._add(data, options?.cache ?? true);
    }
    /**
     * Fetch all channels for the guild
     */
    async fetchAll() {
        const data = await this.client.rest.request('GET', `/guilds/${this.guild.id}/channels`);
        const channels = new Collection_1.Collection();
        for (const channelData of data) {
            const channel = this._add(channelData);
            channels.set(channel.id, channel);
        }
        return channels;
    }
    /**
     * Create a new channel
     */
    async create(options) {
        const body = {
            name: options.name,
            type: options.type ?? ChannelType.GuildText,
        };
        if (options.topic)
            body.topic = options.topic;
        if (options.bitrate)
            body.bitrate = options.bitrate;
        if (options.userLimit)
            body.user_limit = options.userLimit;
        if (options.rateLimitPerUser)
            body.rate_limit_per_user = options.rateLimitPerUser;
        if (options.position !== undefined)
            body.position = options.position;
        if (options.permissionOverwrites)
            body.permission_overwrites = options.permissionOverwrites;
        if (options.parent)
            body.parent_id = options.parent;
        if (options.nsfw !== undefined)
            body.nsfw = options.nsfw;
        const data = await this.client.rest.request('POST', `/guilds/${this.guild.id}/channels`, body);
        return this._add(data);
    }
    /**
     * Delete a channel
     */
    async delete(id, reason) {
        await this.client.rest.request('DELETE', `/channels/${id}`, reason ? { reason } : undefined);
        this.cache.delete(id);
    }
    /**
     * Edit a channel
     */
    async edit(id, data) {
        const body = {};
        if (data.name)
            body.name = data.name;
        if (data.type !== undefined)
            body.type = data.type;
        if (data.position !== undefined)
            body.position = data.position;
        if (data.topic !== undefined)
            body.topic = data.topic;
        if (data.nsfw !== undefined)
            body.nsfw = data.nsfw;
        if (data.rateLimitPerUser !== undefined)
            body.rate_limit_per_user = data.rateLimitPerUser;
        if (data.bitrate !== undefined)
            body.bitrate = data.bitrate;
        if (data.userLimit !== undefined)
            body.user_limit = data.userLimit;
        if (data.permissionOverwrites)
            body.permission_overwrites = data.permissionOverwrites;
        if (data.parent !== undefined)
            body.parent_id = data.parent;
        const result = await this.client.rest.request('PATCH', `/channels/${id}`, body);
        return this._add(result);
    }
    /**
     * Set channel positions
     */
    async setPositions(positions) {
        const body = positions.map(p => ({
            id: p.channel,
            position: p.position,
            parent_id: p.parent,
        }));
        await this.client.rest.request('PATCH', `/guilds/${this.guild.id}/channels`, body);
    }
}
exports.GuildChannelManager = GuildChannelManager;
/**
 * Global channel manager for the client
 */
class ChannelManager extends BaseManager_1.CachedManager {
    constructor(client, iterable) {
        super(client, Object, iterable);
    }
    _add(data, cache = true) {
        const id = data.id;
        const existing = this.cache.get(id);
        if (existing) {
            if (cache)
                Object.assign(existing, data);
            return existing;
        }
        const channel = { ...data, id };
        if (cache)
            this.cache.set(id, channel);
        return channel;
    }
    async fetch(id, options) {
        if (!options?.force) {
            const existing = this.cache.get(id);
            if (existing)
                return existing;
        }
        const data = await this.client.rest.request('GET', `/channels/${id}`);
        return this._add(data, options?.cache ?? true);
    }
}
exports.ChannelManager = ChannelManager;
exports.default = GuildChannelManager;
