// @jubbio/core/dist/managers/RoleManager.js
"use strict";
/**
 * Manager for roles with caching and lazy loading
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoleManager = void 0;
const BaseManager_1 = require("./BaseManager");
const Collection_1 = require("../utils/Collection");
/**
 * Manages roles for a guild
 */
class RoleManager extends BaseManager_1.CachedManager {
    /** The guild this manager belongs to */
    guild;
    constructor(guild, iterable) {
        super(guild.client, Object, iterable);
        this.guild = guild;
    }
    /**
     * Get the @everyone role
     */
    get everyone() {
        return this.cache.get(this.guild.id) ?? null;
    }
    /**
     * Get the highest role
     */
    get highest() {
        return this.cache.reduce((prev, role) => (role.position > (prev?.position ?? -1)) ? role : prev, null);
    }
    /**
     * Get the bot's highest role
     */
    get botRoleFor() {
        return (userId) => {
            const member = this.guild.members?.cache.get(userId);
            if (!member)
                return null;
            return this.cache
                .filter((role) => member.roles?.includes(role.id))
                .sort((a, b) => b.position - a.position)
                .first() ?? null;
        };
    }
    /**
     * Add a role to the cache
     */
    _add(data, cache = true) {
        const id = data.id;
        const existing = this.cache.get(id);
        if (existing) {
            if (cache)
                Object.assign(existing, data);
            return existing;
        }
        const role = {
            id,
            guildId: this.guild.id,
            name: data.name,
            color: data.color ?? 0,
            hoist: data.hoist ?? false,
            position: data.position ?? 0,
            permissions: data.permissions ?? '0',
            managed: data.managed ?? false,
            mentionable: data.mentionable ?? false,
            icon: data.icon ?? null,
            unicodeEmoji: data.unicode_emoji ?? null,
            tags: data.tags ?? null,
            get hexColor() {
                return `#${this.color.toString(16).padStart(6, '0')}`;
            },
            get createdAt() {
                const timestamp = BigInt(this.id) >> 22n;
                return new Date(Number(timestamp) + 1640995200000); // Jubbio epoch
            },
            toString() {
                return this.id === this.guildId ? '@everyone' : `<@&${this.id}>`;
            },
            comparePositionTo(role) {
                return this.position - (role?.position ?? 0);
            },
            async edit(data) {
                return this.guild.roles.edit(this.id, data);
            },
            async delete(reason) {
                return this.guild.roles.delete(this.id, reason);
            },
            async setPosition(position) {
                return this.guild.roles.setPositions([{ role: this.id, position }]);
            },
        };
        if (cache)
            this.cache.set(id, role);
        return role;
    }
    /**
     * Fetch a role from the API
     */
    async fetch(id, options) {
        if (!options?.force) {
            const existing = this.cache.get(id);
            if (existing)
                return existing;
        }
        // Roles are fetched as part of guild, so fetch all
        const roles = await this.fetchAll();
        return roles.get(id) ?? null;
    }
    /**
     * Fetch all roles for the guild
     */
    async fetchAll() {
        const data = await this.client.rest.request('GET', `/guilds/${this.guild.id}/roles`);
        const roles = new Collection_1.Collection();
        for (const roleData of data) {
            const role = this._add(roleData);
            roles.set(role.id, role);
        }
        return roles;
    }
    /**
     * Create a new role
     */
    async create(options) {
        const body = {};
        if (options?.name)
            body.name = options.name;
        if (options?.color !== undefined) {
            body.color = typeof options.color === 'string'
                ? parseInt(options.color.replace('#', ''), 16)
                : options.color;
        }
        if (options?.hoist !== undefined)
            body.hoist = options.hoist;
        if (options?.permissions !== undefined)
            body.permissions = String(options.permissions);
        if (options?.mentionable !== undefined)
            body.mentionable = options.mentionable;
        if (options?.icon)
            body.icon = options.icon;
        if (options?.unicodeEmoji)
            body.unicode_emoji = options.unicodeEmoji;
        const data = await this.client.rest.request('POST', `/guilds/${this.guild.id}/roles`, body);
        const role = this._add(data);
        // Set position if specified
        if (options?.position !== undefined) {
            await this.setPositions([{ role: role.id, position: options.position }]);
        }
        return role;
    }
    /**
     * Delete a role
     */
    async delete(id, reason) {
        await this.client.rest.request('DELETE', `/guilds/${this.guild.id}/roles/${id}`, reason ? { reason } : undefined);
        this.cache.delete(id);
    }
    /**
     * Edit a role
     */
    async edit(id, data) {
        const body = {};
        if (data.name !== undefined)
            body.name = data.name;
        if (data.color !== undefined) {
            body.color = typeof data.color === 'string'
                ? parseInt(data.color.replace('#', ''), 16)
                : data.color;
        }
        if (data.hoist !== undefined)
            body.hoist = data.hoist;
        if (data.permissions !== undefined)
            body.permissions = String(data.permissions);
        if (data.mentionable !== undefined)
            body.mentionable = data.mentionable;
        if (data.icon !== undefined)
            body.icon = data.icon;
        if (data.unicodeEmoji !== undefined)
            body.unicode_emoji = data.unicodeEmoji;
        const result = await this.client.rest.request('PATCH', `/guilds/${this.guild.id}/roles/${id}`, body);
        return this._add(result);
    }
    /**
     * Set role positions
     */
    async setPositions(positions) {
        const body = positions.map(p => ({ id: p.role, position: p.position }));
        const data = await this.client.rest.request('PATCH', `/guilds/${this.guild.id}/roles`, body);
        const roles = new Collection_1.Collection();
        for (const roleData of data) {
            const role = this._add(roleData);
            roles.set(role.id, role);
        }
        return roles;
    }
    /**
     * Compare two roles by position
     */
    comparePositions(role1, role2) {
        const r1 = typeof role1 === 'string' ? this.cache.get(role1) : role1;
        const r2 = typeof role2 === 'string' ? this.cache.get(role2) : role2;
        return (r1?.position ?? 0) - (r2?.position ?? 0);
    }
}
exports.RoleManager = RoleManager;
exports.default = RoleManager;
