// @jubbio/voice/dist/AudioResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioResource = void 0;
exports.createAudioResource = createAudioResource;
exports.createAudioResourceFromUrl = createAudioResourceFromUrl;
exports.probeAudioInfo = probeAudioInfo;
const child_process_1 = require("child_process");
const enums_1 = require("./enums");
/**
 * Represents an audio resource that can be played
 */
class AudioResource {
    /** Metadata attached to this resource */
    metadata;
    /** Whether playback has started */
    started = false;
    /** Whether playback has ended */
    ended = false;
    /** The input source (URL or file path) */
    inputSource;
    /** Stream type */
    streamType;
    /** Volume (0-1) */
    volume = 1;
    constructor(input, options = {}) {
        this.metadata = options.metadata;
        this.streamType = options.inputType || enums_1.StreamType.Arbitrary;
        if (typeof input === 'string') {
            this.inputSource = input;
        }
        else {
            // For streams, we'd need to handle differently
            // For now, throw an error
            throw new Error('Stream input not yet supported. Use URL or file path.');
        }
    }
    /**
     * Get the input source for FFmpeg
     * @internal
     */
    getInputSource() {
        return this.inputSource;
    }
    /**
     * Set the volume (0-1)
     */
    setVolume(volume) {
        this.volume = Math.max(0, Math.min(1, volume));
    }
    /**
     * Get the current volume
     */
    getVolume() {
        return this.volume;
    }
}
exports.AudioResource = AudioResource;
/**
 * Create an audio resource from various inputs
 */
function createAudioResource(input, options) {
    return new AudioResource(input, options);
}
/**
 * Create an audio resource from a YouTube/streaming URL
 * Stores the original URL - extraction happens at playback time
 */
function createAudioResourceFromUrl(url, options = {}) {
    // Don't extract stream URL here - just store the original URL
    // The AudioPlayer will use yt-dlp at playback time
    return new AudioResource(url, options);
}
/**
 * Check if URL is a streaming service URL
 */
function isStreamingUrl(url) {
    const streamingDomains = [
        'youtube.com',
        'youtu.be',
        'soundcloud.com',
        'spotify.com',
        'twitch.tv',
        'vimeo.com'
    ];
    return streamingDomains.some(domain => url.includes(domain));
}
/**
 * Check if input is a valid URL
 */
function isValidUrl(input) {
    try {
        new URL(input);
        return true;
    }
    catch {
        return input.startsWith('http://') || input.startsWith('https://');
    }
}
/**
 * Probe audio info from a URL or search query
 * If input is not a URL, it will search YouTube
 */
async function probeAudioInfo(input, ytDlpPath) {
    return new Promise((resolve, reject) => {
        const isWindows = process.platform === 'win32';
        const defaultYtDlpPath = isWindows ? 'yt-dlp' : '~/.local/bin/yt-dlp';
        const ytdlpBin = ytDlpPath || defaultYtDlpPath;
        // If not a valid URL, treat as YouTube search
        let searchQuery = input;
        if (!isValidUrl(input)) {
            searchQuery = `ytsearch1:${input}`;
        }
        let ytdlp;
        if (isWindows) {
            // Windows: escape special characters and use shell
            // Replace double quotes with escaped version for cmd
            const escapedQuery = searchQuery.replace(/"/g, '\\"');
            const cmd = `${ytdlpBin} --no-playlist --no-warnings -j "${escapedQuery}"`;
            console.log('[probeAudioInfo] Running:', cmd);
            ytdlp = (0, child_process_1.spawn)(cmd, [], { shell: true });
        }
        else {
            // Unix: use bash -c with quoted string
            ytdlp = (0, child_process_1.spawn)('bash', [
                '-c',
                `${ytdlpBin} --no-playlist --no-warnings -j "${searchQuery}"`
            ]);
        }
        let stdout = '';
        let stderr = '';
        ytdlp.stdout?.on('data', (data) => {
            stdout += data.toString();
        });
        ytdlp.stderr?.on('data', (data) => {
            stderr += data.toString();
        });
        ytdlp.on('close', (code) => {
            if (code !== 0) {
                reject(new Error(`yt-dlp failed (code ${code}): ${stderr || 'Unknown error'}`));
                return;
            }
            if (!stdout.trim()) {
                reject(new Error(`yt-dlp returned empty response. stderr: ${stderr}`));
                return;
            }
            try {
                const info = JSON.parse(stdout);
                resolve({
                    title: info.title || 'Unknown',
                    duration: info.duration || 0,
                    thumbnail: info.thumbnail,
                    url: info.webpage_url || info.url || input
                });
            }
            catch (e) {
                reject(new Error(`Failed to parse yt-dlp output: ${stdout.substring(0, 200)}`));
            }
        });
        ytdlp.on('error', (err) => {
            reject(new Error(`Failed to probe audio info: ${err.message}`));
        });
        // Timeout after 30 seconds
        setTimeout(() => {
            ytdlp.kill();
            reject(new Error('Timeout waiting for audio info'));
        }, 30000);
    });
}
