// @jubbio/voice/dist/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamType = exports.AudioPlayerStatus = exports.VoiceConnectionStatus = void 0;
/**
 * Voice connection status
 */
var VoiceConnectionStatus;
(function (VoiceConnectionStatus) {
    /** Connection is being established */
    VoiceConnectionStatus["Connecting"] = "connecting";
    /** Connection is ready to use */
    VoiceConnectionStatus["Ready"] = "ready";
    /** Connection is disconnected */
    VoiceConnectionStatus["Disconnected"] = "disconnected";
    /** Connection is destroyed */
    VoiceConnectionStatus["Destroyed"] = "destroyed";
    /** Connection is signalling (exchanging info with server) */
    VoiceConnectionStatus["Signalling"] = "signalling";
})(VoiceConnectionStatus || (exports.VoiceConnectionStatus = VoiceConnectionStatus = {}));
/**
 * Audio player status
 */
var AudioPlayerStatus;
(function (AudioPlayerStatus) {
    /** Player is idle (not playing anything) */
    AudioPlayerStatus["Idle"] = "idle";
    /** Player is buffering audio */
    AudioPlayerStatus["Buffering"] = "buffering";
    /** Player is playing audio */
    AudioPlayerStatus["Playing"] = "playing";
    /** Player is paused */
    AudioPlayerStatus["Paused"] = "paused";
    /** Player is auto-paused (no subscribers) */
    AudioPlayerStatus["AutoPaused"] = "autopaused";
})(AudioPlayerStatus || (exports.AudioPlayerStatus = AudioPlayerStatus = {}));
/**
 * Audio resource type
 */
var StreamType;
(function (StreamType) {
    /** Arbitrary audio stream */
    StreamType["Arbitrary"] = "arbitrary";
    /** Raw PCM audio */
    StreamType["Raw"] = "raw";
    /** Opus encoded audio */
    StreamType["Opus"] = "opus";
    /** OGG/Opus container */
    StreamType["OggOpus"] = "ogg/opus";
    /** WebM/Opus container */
    StreamType["WebmOpus"] = "webm/opus";
})(StreamType || (exports.StreamType = StreamType = {}));
