// @jubbio/voice/dist/VoiceConnection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceConnection = void 0;
exports.joinVoiceChannel = joinVoiceChannel;
exports.getVoiceConnection = getVoiceConnection;
const events_1 = require("events");
const rtc_node_1 = require("@livekit/rtc-node");
const enums_1 = require("./enums");
/**
 * Represents a voice connection to a channel
 */
class VoiceConnection extends events_1.EventEmitter {
    /** Current connection state */
    state = { status: enums_1.VoiceConnectionStatus.Connecting };
    /** The channel ID this connection is for */
    channelId;
    /** The guild ID this connection is for */
    guildId;
    /** LiveKit room instance */
    room = null;
    /** LiveKit connection info */
    livekitEndpoint = null;
    livekitToken = null;
    livekitRoomName = null;
    /** Subscribed audio player */
    subscribedPlayer = null;
    /** Gateway adapter methods */
    adapterMethods;
    /** Adapter implementer (for sending payloads) */
    adapter = null;
    constructor(options) {
        super();
        this.channelId = options.channelId;
        this.guildId = options.guildId;
        // Create adapter methods that will receive gateway events
        this.adapterMethods = {
            onVoiceServerUpdate: (data) => this.handleVoiceServerUpdate(data),
            onVoiceStateUpdate: (data) => this.handleVoiceStateUpdate(data),
            destroy: () => this.destroy()
        };
        // Get adapter from creator
        this.adapter = options.adapterCreator(this.adapterMethods);
        // Send voice state update to join channel
        this.sendVoiceStateUpdate(options.channelId, options.selfMute, options.selfDeaf);
    }
    /**
     * Subscribe an audio player to this connection
     */
    subscribe(player) {
        if (this.subscribedPlayer) {
            this.subscribedPlayer.unsubscribe(this);
        }
        this.subscribedPlayer = player;
        player.subscribe(this);
    }
    /**
     * Unsubscribe the current audio player
     */
    unsubscribe() {
        if (this.subscribedPlayer) {
            this.subscribedPlayer.unsubscribe(this);
            this.subscribedPlayer = null;
        }
    }
    /**
     * Get the LiveKit room (for audio player to publish tracks)
     */
    getRoom() {
        return this.room;
    }
    /**
     * Disconnect from the voice channel
     */
    disconnect() {
        this.sendVoiceStateUpdate(null);
        return true;
    }
    /**
     * Destroy the connection completely
     */
    destroy() {
        this.unsubscribe();
        this.disconnectFromLiveKit();
        this.adapter?.destroy();
        this.setState({ status: enums_1.VoiceConnectionStatus.Destroyed });
    }
    /**
     * Rejoin the voice channel (after disconnect)
     */
    rejoin() {
        this.sendVoiceStateUpdate(this.channelId);
        return true;
    }
    sendVoiceStateUpdate(channelId, selfMute = false, selfDeaf = false) {
        if (!this.adapter)
            return;
        this.adapter.sendPayload({
            op: 4, // VOICE_STATE_UPDATE
            d: {
                guild_id: this.guildId,
                channel_id: channelId,
                self_mute: selfMute,
                self_deaf: selfDeaf
            }
        });
    }
    handleVoiceServerUpdate(data) {
        this.livekitEndpoint = data.endpoint;
        this.livekitToken = data.token;
        this.livekitRoomName = data.room;
        this.setState({ status: enums_1.VoiceConnectionStatus.Signalling });
        this.connectToLiveKit();
    }
    handleVoiceStateUpdate(data) {
        if (data.channel_id === null) {
            // Disconnected
            this.disconnectFromLiveKit();
            this.setState({ status: enums_1.VoiceConnectionStatus.Disconnected });
        }
    }
    async connectToLiveKit() {
        if (!this.livekitEndpoint || !this.livekitToken) {
            this.emit('error', new Error('Missing LiveKit connection info'));
            return;
        }
        try {
            // Disconnect existing room if any
            await this.disconnectFromLiveKit();
            this.room = new rtc_node_1.Room();
            this.room.on(rtc_node_1.RoomEvent.Disconnected, () => {
                this.setState({ status: enums_1.VoiceConnectionStatus.Disconnected });
            });
            await this.room.connect(this.livekitEndpoint, this.livekitToken);
            this.setState({ status: enums_1.VoiceConnectionStatus.Ready });
            // Notify subscribed player that connection is ready
            if (this.subscribedPlayer) {
                this.subscribedPlayer.onConnectionReady(this);
            }
        }
        catch (error) {
            this.emit('error', error);
            this.setState({ status: enums_1.VoiceConnectionStatus.Disconnected });
        }
    }
    async disconnectFromLiveKit() {
        if (this.room) {
            try {
                await this.room.disconnect();
            }
            catch (e) {
                // Ignore disconnect errors
            }
            this.room = null;
        }
    }
    setState(newState) {
        const oldState = this.state;
        this.state = newState;
        this.emit('stateChange', oldState, newState);
    }
}
exports.VoiceConnection = VoiceConnection;
// Store active connections
const connections = new Map();
/**
 * Join a voice channel - main entry point
 */
function joinVoiceChannel(options) {
    const key = `${options.guildId}:${options.channelId}`;
    // Destroy existing connection if any
    const existing = connections.get(key);
    if (existing) {
        existing.destroy();
        connections.delete(key);
    }
    const connection = new VoiceConnection(options);
    connections.set(key, connection);
    return connection;
}
/**
 * Get an existing voice connection
 */
function getVoiceConnection(guildId) {
    for (const [key, connection] of connections) {
        if (key.startsWith(`${guildId}:`)) {
            return connection;
        }
    }
    return undefined;
}
