// @jubbio/core/dist/structures/Channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DMChannel = exports.VoiceChannel = exports.TextChannel = exports.BaseChannel = void 0;
exports.createChannel = createChannel;
const enums_1 = require("../enums");
const Message_1 = require("./Message");
const Collection_1 = require("./Collection");
const Collector_1 = require("../utils/Collector");
const EmbedBuilder_1 = require("../builders/EmbedBuilder");
/** Resolve EmbedBuilder instances to plain API objects */
function resolveEmbeds(embeds) {
    if (!embeds)
        return undefined;
    return embeds.map(e => e instanceof EmbedBuilder_1.EmbedBuilder ? e.toJSON() : e);
}
/**
 * Base channel class
 */
class BaseChannel {
    /** Reference to the client */
    client;
    /** Channel ID */
    id;
    /** Channel type */
    type;
    constructor(client, data) {
        this.client = client;
        this.id = data.id;
        this.type = data.type;
    }
    /**
     * Check if this is a text-based channel
     */
    isTextBased() {
        return [
            enums_1.ChannelType.GuildText,
            enums_1.ChannelType.DM,
            enums_1.ChannelType.GroupDM,
            enums_1.ChannelType.GuildAnnouncement
        ].includes(this.type);
    }
    /**
     * Check if this is a voice-based channel
     */
    isVoiceBased() {
        return [
            enums_1.ChannelType.GuildVoice,
            enums_1.ChannelType.GuildStageVoice
        ].includes(this.type);
    }
    /**
     * Convert to string (mention format)
     */
    toString() {
        return `<#${this.id}>`;
    }
}
exports.BaseChannel = BaseChannel;
/**
 * Text channel
 */
class TextChannel extends BaseChannel {
    /** Guild ID */
    guildId;
    /** Channel name */
    name;
    /** Channel topic */
    topic;
    /** Channel position */
    position;
    /** Parent category ID */
    parentId;
    constructor(client, data) {
        super(client, data);
        this.guildId = data.guild_id;
        this.name = data.name;
        this.topic = data.topic;
        this.position = data.position;
        this.parentId = data.parent_id;
    }
    /**
     * Send a message to this channel
     */
    async send(options) {
        const content = typeof options === 'string' ? options : options.content;
        const embeds = typeof options === 'string' ? undefined : resolveEmbeds(options.embeds);
        const components = typeof options === 'string' ? undefined : options.components;
        const data = await this.client.rest.createMessage(this.guildId || '', this.id, {
            content,
            embeds,
            components,
        });
        return new Message_1.Message(this.client, data);
    }
    /**
     * Create a message collector
     */
    createMessageCollector(options) {
        return new Collector_1.MessageCollector(this.client, this.id, options);
    }
    /**
     * Await messages in this channel
     */
    awaitMessages(options) {
        return new Promise((resolve, reject) => {
            const collector = this.createMessageCollector(options);
            collector.once('end', (collected, reason) => {
                if (options?.errors?.includes(reason)) {
                    reject(collected);
                }
                else {
                    resolve(collected);
                }
            });
        });
    }
    /**
     * Bulk delete messages
     */
    async bulkDelete(messages, filterOld = true) {
        let messageIds;
        if (typeof messages === 'number') {
            // Fetch messages first
            const fetched = await this.client.rest.getMessages(this.guildId || '', this.id, { limit: messages });
            messageIds = fetched.map(m => m.id);
        }
        else if (messages instanceof Collection_1.Collection) {
            messageIds = [...messages.keys()];
        }
        else {
            messageIds = messages;
        }
        // Filter old messages (older than 14 days)
        if (filterOld) {
            const twoWeeksAgo = Date.now() - 14 * 24 * 60 * 60 * 1000;
            messageIds = messageIds.filter(id => {
                // Extract timestamp from snowflake
                const timestamp = Number(BigInt(id) >> 22n) + 1420070400000;
                return timestamp > twoWeeksAgo;
            });
        }
        if (messageIds.length === 0) {
            return new Collection_1.Collection();
        }
        if (messageIds.length === 1) {
            await this.client.rest.deleteMessage(this.guildId || '', this.id, messageIds[0]);
        }
        else {
            await this.client.rest.bulkDeleteMessages(this.guildId || '', this.id, messageIds);
        }
        // Return deleted messages as collection
        const deleted = new Collection_1.Collection();
        // Note: We don't have the actual message objects here
        return deleted;
    }
}
exports.TextChannel = TextChannel;
/**
 * Voice channel
 */
class VoiceChannel extends BaseChannel {
    /** Guild ID */
    guildId;
    /** Channel name */
    name;
    /** Channel position */
    position;
    /** Parent category ID */
    parentId;
    /** User limit */
    userLimit;
    /** Bitrate */
    bitrate;
    constructor(client, data) {
        super(client, data);
        this.guildId = data.guild_id;
        this.name = data.name;
        this.position = data.position;
        this.parentId = data.parent_id;
        this.userLimit = data.user_limit;
        this.bitrate = data.bitrate;
    }
    /**
     * Check if the channel is joinable
     */
    get joinable() {
        // TODO: Check permissions
        return true;
    }
}
exports.VoiceChannel = VoiceChannel;
/**
 * DM channel
 */
class DMChannel extends BaseChannel {
    /** Recipient user ID */
    recipientId;
    constructor(client, data) {
        super(client, data);
        this.recipientId = data.recipient_id;
    }
    /**
     * Send a message to this DM
     */
    async send(options) {
        const content = typeof options === 'string' ? options : options.content;
        const embeds = typeof options === 'string' ? undefined : resolveEmbeds(options.embeds);
        const data = await this.client.rest.createDMMessage(this.id, {
            content,
            embeds
        });
        return new Message_1.Message(this.client, data);
    }
}
exports.DMChannel = DMChannel;
/**
 * Create appropriate channel class based on type
 */
function createChannel(client, data) {
    switch (data.type) {
        case enums_1.ChannelType.GuildText:
        case enums_1.ChannelType.GuildAnnouncement:
            return new TextChannel(client, data);
        case enums_1.ChannelType.GuildVoice:
        case enums_1.ChannelType.GuildStageVoice:
            return new VoiceChannel(client, data);
        case enums_1.ChannelType.DM:
            return new DMChannel(client, data);
        default:
            return new BaseChannel(client, data);
    }
}
