// @jubbio/core/dist/structures/Collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collection = void 0;
/**
 * Extended Map with utility methods
 */
class Collection extends Map {
    /**
     * Get the first value in the collection
     */
    first() {
        return this.values().next().value;
    }
    /**
     * Get the last value in the collection
     */
    last() {
        const arr = [...this.values()];
        return arr[arr.length - 1];
    }
    /**
     * Get a random value from the collection
     */
    random() {
        const arr = [...this.values()];
        return arr[Math.floor(Math.random() * arr.length)];
    }
    /**
     * Find a value matching a predicate
     */
    find(fn) {
        for (const [key, val] of this) {
            if (fn(val, key, this))
                return val;
        }
        return undefined;
    }
    /**
     * Filter values matching a predicate
     */
    filter(fn) {
        const results = new Collection();
        for (const [key, val] of this) {
            if (fn(val, key, this))
                results.set(key, val);
        }
        return results;
    }
    /**
     * Map values to a new array
     */
    map(fn) {
        const results = [];
        for (const [key, val] of this) {
            results.push(fn(val, key, this));
        }
        return results;
    }
    /**
     * Check if some values match a predicate
     */
    some(fn) {
        for (const [key, val] of this) {
            if (fn(val, key, this))
                return true;
        }
        return false;
    }
    /**
     * Check if every value matches a predicate
     */
    every(fn) {
        for (const [key, val] of this) {
            if (!fn(val, key, this))
                return false;
        }
        return true;
    }
    /**
     * Reduce the collection to a single value
     */
    reduce(fn, initialValue) {
        let accumulator = initialValue;
        for (const [key, val] of this) {
            accumulator = fn(accumulator, val, key, this);
        }
        return accumulator;
    }
    /**
     * Convert to array
     */
    toArray() {
        return [...this.values()];
    }
    /**
     * Clone the collection
     */
    clone() {
        return new Collection(this);
    }
    /**
     * Concat with another collection
     */
    concat(...collections) {
        const newColl = this.clone();
        for (const coll of collections) {
            for (const [key, val] of coll) {
                newColl.set(key, val);
            }
        }
        return newColl;
    }
}
exports.Collection = Collection;
