// @jubbio/core/dist/structures/GuildMember.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildMember = exports.Permissions = exports.PermissionFlags = void 0;
const User_1 = require("./User");
/**
 * Permission flags - permission bits
 * Must match community-service/internal/guild/models/permissions.go (truth of source)
 */
exports.PermissionFlags = {
    // General (0-9)
    CreateInstantInvite: 1n << 0n,
    KickMembers: 1n << 1n,
    BanMembers: 1n << 2n,
    Administrator: 1n << 3n,
    ManageChannels: 1n << 4n,
    ManageGuild: 1n << 5n,
    AddReactions: 1n << 6n,
    ViewAuditLog: 1n << 7n,
    ViewGuildInsights: 1n << 8n,
    Stream: 1n << 9n,
    ViewChannel: 1n << 10n,
    // Messages (11-19)
    SendMessages: 1n << 11n,
    SendTTSMessages: 1n << 12n,
    ManageMessages: 1n << 13n,
    EmbedLinks: 1n << 14n,
    AttachFiles: 1n << 15n,
    ReadMessageHistory: 1n << 16n,
    UseExternalEmojis: 1n << 17n,
    UseSlashCommands: 1n << 18n,
    MentionEveryone: 1n << 19n,
    // Voice (20-27)
    Connect: 1n << 20n,
    Speak: 1n << 21n,
    MuteMembers: 1n << 22n,
    DeafenMembers: 1n << 23n,
    MoveMembers: 1n << 24n,
    UseVAD: 1n << 25n,
    ChangeCodec: 1n << 26n,
    AudioQualityAdmin: 1n << 27n,
    // Video and Screen Sharing (28-37)
    VideoCall: 1n << 28n,
    ShareScreen: 1n << 29n,
    ShareCamera: 1n << 30n,
    ControlQuality: 1n << 31n,
    RequestToSpeak: 1n << 32n,
    ManageEvents: 1n << 33n,
    AddMembers: 1n << 34n,
    RemoveMembers: 1n << 35n,
    ChangeGroupIcon: 1n << 36n,
    ChangeDMSettings: 1n << 37n,
    // Advanced (38-43)
    ManageGroup: 1n << 38n,
    UseActivities: 1n << 39n,
    ModerateMembers: 1n << 40n,
    ManageRoles: 1n << 41n,
    ManageEmojis: 1n << 42n,
    PrioritySpeaker: 1n << 43n,
};
/**
 * Permissions helper class
 */
class Permissions {
    bitfield;
    isOwner;
    constructor(bits = 0n, isOwner = false) {
        if (typeof bits === 'string') {
            this.bitfield = BigInt(bits);
        }
        else {
            this.bitfield = BigInt(bits);
        }
        this.isOwner = isOwner;
    }
    /**
     * Check if has a permission
     */
    has(permission) {
        // Owner has all permissions
        if (this.isOwner)
            return true;
        // Administrator has all permissions
        if ((this.bitfield & exports.PermissionFlags.Administrator) === exports.PermissionFlags.Administrator) {
            return true;
        }
        let bit;
        if (typeof permission === 'string') {
            bit = exports.PermissionFlags[permission] || 0n;
        }
        else {
            bit = permission;
        }
        return (this.bitfield & bit) === bit;
    }
    /**
     * Get the raw bitfield
     */
    get bits() {
        return this.bitfield;
    }
    /**
     * Convert to array of permission names
     */
    toArray() {
        const result = [];
        for (const [name, bit] of Object.entries(exports.PermissionFlags)) {
            if ((this.bitfield & bit) === bit) {
                result.push(name);
            }
        }
        return result;
    }
}
exports.Permissions = Permissions;
/**
 * Represents a guild member
 */
class GuildMember {
    /** Reference to the client */
    client;
    /** Reference to the guild */
    guild;
    /** The user this member represents */
    user;
    /** Member's nickname */
    nickname;
    /** Member's guild avatar */
    avatar;
    /** Role IDs */
    roles;
    /** Join timestamp */
    joinedTimestamp;
    /** Voice state */
    voice;
    /** Member permissions */
    permissions;
    constructor(client, guild, data) {
        this.client = client;
        this.guild = guild;
        this.user = data.user ? new User_1.User(data.user) : new User_1.User({ id: '0', username: 'Unknown' });
        this.nickname = data.nick;
        this.avatar = data.avatar;
        this.roles = data.roles;
        this.joinedTimestamp = new Date(data.joined_at).getTime();
        this.voice = {
            channelId: data.voice?.channel_id,
            selfMute: data.voice?.self_mute ?? false,
            selfDeaf: data.voice?.self_deaf ?? false
        };
        // Check if user is guild owner
        const isOwner = guild.ownerId === this.user.id;
        // Parse permissions from interaction data
        this.permissions = new Permissions(data.permissions || '0', isOwner);
    }
    /**
     * Get the member's ID
     */
    get id() {
        return this.user.id;
    }
    /**
     * Get the display name (nickname or username)
     */
    get displayName() {
        return this.nickname || this.user.displayName || this.user.username;
    }
    /**
     * Get the join date
     */
    get joinedAt() {
        return new Date(this.joinedTimestamp);
    }
    /**
     * Check if member is in a voice channel
     */
    get inVoiceChannel() {
        return !!this.voice.channelId;
    }
    /**
     * Get the member's avatar URL
     */
    avatarURL() {
        return this.avatar || null;
    }
    /**
     * Get the display avatar URL (member avatar or user avatar)
     */
    displayAvatarURL() {
        return this.avatar || this.user.displayAvatarURL();
    }
    /**
     * Convert to string (mention format)
     */
    toString() {
        return `<@${this.id}>`;
    }
    /**
     * Update member data
     */
    _patch(data) {
        if (data.nick !== undefined)
            this.nickname = data.nick;
        if (data.avatar !== undefined)
            this.avatar = data.avatar;
        if (data.roles !== undefined)
            this.roles = data.roles;
        if (data.voice !== undefined) {
            this.voice = {
                channelId: data.voice.channel_id,
                selfMute: data.voice.self_mute ?? false,
                selfDeaf: data.voice.self_deaf ?? false
            };
        }
    }
}
exports.GuildMember = GuildMember;
