// @jubbio/core/dist/structures/Guild.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Guild = void 0;
const Collection_1 = require("./Collection");
const GuildMember_1 = require("./GuildMember");
/**
 * Represents a guild
 */
class Guild {
    /** Reference to the client */
    client;
    /** Guild ID */
    id;
    /** Guild name */
    name;
    /** Guild icon URL */
    icon;
    /** Owner ID */
    ownerId;
    /** Whether the guild is unavailable */
    unavailable;
    /** Cached members */
    members;
    /** Cached channels */
    channels;
    constructor(client, data) {
        this.client = client;
        this.id = data.id;
        this.name = data.name;
        this.icon = data.icon;
        this.ownerId = data.owner_id;
        this.unavailable = data.unavailable ?? false;
        this.members = new Collection_1.Collection();
        this.channels = new Collection_1.Collection();
    }
    /**
     * Get the guild icon URL
     */
    iconURL(options) {
        if (!this.icon)
            return null;
        return this.icon;
    }
    /**
     * Get the voice adapter creator for @jubbio/voice
     */
    get voiceAdapterCreator() {
        return this.client.voice.adapters.get(this.id);
    }
    /**
     * Fetch a member by ID
     */
    async fetchMember(userId) {
        // Check cache first
        const cached = this.members.get(userId);
        if (cached)
            return cached;
        // TODO: Fetch from API
        return null;
    }
    /**
     * Convert to string
     */
    toString() {
        return this.name;
    }
    /**
     * Update guild data
     */
    _patch(data) {
        if (data.name !== undefined)
            this.name = data.name;
        if (data.icon !== undefined)
            this.icon = data.icon;
        if (data.owner_id !== undefined)
            this.ownerId = data.owner_id;
        if (data.unavailable !== undefined)
            this.unavailable = data.unavailable;
    }
    /**
     * Add a member to cache
     */
    _addMember(data) {
        const member = new GuildMember_1.GuildMember(this.client, this, data);
        this.members.set(member.id, member);
        return member;
    }
}
exports.Guild = Guild;
