// @jubbio/core/dist/structures/Message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const User_1 = require("./User");
const Collector_1 = require("../utils/Collector");
const EmbedBuilder_1 = require("../builders/EmbedBuilder");
/** Resolve EmbedBuilder instances to plain API objects */
function resolveEmbeds(embeds) {
    if (!embeds)
        return undefined;
    return embeds.map(e => e instanceof EmbedBuilder_1.EmbedBuilder ? e.toJSON() : e);
}
/**
 * Represents a message
 */
class Message {
    /** Reference to the client */
    client;
    /** Message ID */
    id;
    /** Channel ID */
    channelId;
    /** Guild ID (if in a guild) */
    guildId;
    /** Message author */
    author;
    /** Message content */
    content;
    /** Message timestamp */
    createdTimestamp;
    /** Edit timestamp */
    editedTimestamp;
    /** Attachments */
    attachments;
    /** Embeds */
    embeds;
    /** Mentions in the message */
    mentions;
    /** User ID (from backend) */
    user_id;
    constructor(client, data) {
        this.client = client;
        this.id = data.id;
        this.channelId = data.channel_id;
        this.guildId = data.guild_id;
        this.author = new User_1.User(data.author);
        this.content = data.content ?? '';
        // Handle different timestamp formats from backend
        const timestamp = data.timestamp || data.created_at;
        this.createdTimestamp = timestamp ? new Date(timestamp).getTime() : Date.now();
        const editedTimestamp = data.edited_timestamp || data.updated_at;
        this.editedTimestamp = editedTimestamp ? new Date(editedTimestamp).getTime() : undefined;
        this.attachments = data.attachments || [];
        this.embeds = data.embeds || [];
        this.mentions = data.mentions || {};
        this.user_id = data.user_id;
    }
    /**
     * Get the creation date
     */
    get createdAt() {
        return new Date(this.createdTimestamp);
    }
    /**
     * Get the edit date
     */
    get editedAt() {
        return this.editedTimestamp ? new Date(this.editedTimestamp) : null;
    }
    /**
     * Reply to this message
     */
    async reply(options) {
        const content = typeof options === 'string' ? options : options.content;
        const embeds = typeof options === 'string' ? undefined : resolveEmbeds(options.embeds);
        const components = typeof options === 'string' ? undefined : options.components;
        const data = await this.client.rest.createMessage(this.guildId || '', this.channelId, {
            content,
            embeds,
            components,
            message_reference: { message_id: this.id }
        });
        return new Message(this.client, data);
    }
    /**
     * Edit this message (only if author is the bot)
     */
    async edit(options) {
        const content = typeof options === 'string' ? options : options.content;
        const embeds = typeof options === 'string' ? undefined : resolveEmbeds(options.embeds);
        const components = typeof options === 'string' ? undefined : options.components;
        const data = await this.client.rest.editMessage(this.guildId || '', this.channelId, this.id, {
            content,
            embeds,
            components,
        });
        return new Message(this.client, data);
    }
    /**
     * Delete this message
     */
    async delete() {
        await this.client.rest.deleteMessage(this.guildId || '', this.channelId, this.id);
    }
    /**
     * React to this message
     */
    async react(emoji) {
        await this.client.rest.addReaction(this.guildId || '', this.channelId, this.id, emoji);
    }
    /**
     * Pin this message
     */
    async pin() {
        await this.client.rest.pinMessage(this.guildId || '', this.channelId, this.id);
    }
    /**
     * Unpin this message
     */
    async unpin() {
        await this.client.rest.unpinMessage(this.guildId || '', this.channelId, this.id);
    }
    /**
     * Create a component interaction collector on this message
     */
    createMessageComponentCollector(options) {
        return new Collector_1.InteractionCollector(this.client, {
            ...options,
            messageId: this.id,
            channelId: this.channelId,
            guildId: this.guildId,
        });
    }
    /**
     * Await component interactions on this message
     */
    awaitMessageComponent(options) {
        return new Promise((resolve, reject) => {
            const collector = this.createMessageComponentCollector({ ...options, max: 1 });
            collector.once('end', (collected, reason) => {
                const first = collected.first();
                if (first) {
                    resolve(first);
                }
                else {
                    reject(new Error(reason || 'No interaction received'));
                }
            });
        });
    }
    /**
     * Convert to string
     */
    toString() {
        return this.content;
    }
}
exports.Message = Message;
