// @jubbio/core/dist/structures/User.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
/**
 * Represents a user
 */
class User {
    /** User ID */
    id;
    /** Username */
    username;
    /** Display name */
    displayName;
    /** Avatar URL */
    avatarURL;
    /** Whether this is a bot */
    bot;
    constructor(data) {
        // Handle both string and number IDs (snowflake precision issue)
        this.id = String(data.id);
        this.username = data.username || `User_${this.id}`;
        this.displayName = data.display_name;
        this.avatarURL = data.avatar_url;
        this.bot = data.bot ?? false;
    }
    /**
     * Get the user's tag (username)
     */
    get tag() {
        return this.username;
    }
    /**
     * Get the default avatar URL
     */
    get defaultAvatarURL() {
        return `https://cdn.jubbio.com/embed/avatars/${parseInt(this.id) % 5}.png`;
    }
    /**
     * Get the display avatar URL
     */
    displayAvatarURL() {
        return this.avatarURL || this.defaultAvatarURL;
    }
    /**
     * Convert to string (mention format)
     */
    toString() {
        return `<@${this.id}>`;
    }
    /**
     * Update user data
     */
    _patch(data) {
        if (data.username !== undefined)
            this.username = data.username;
        if (data.display_name !== undefined)
            this.displayName = data.display_name;
        if (data.avatar_url !== undefined)
            this.avatarURL = data.avatar_url;
        if (data.bot !== undefined)
            this.bot = data.bot;
    }
}
exports.User = User;
